// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/flyout_menu/flyout_menu.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "FlyoutMenu", {
    enumerable: true,
    get: function() {
        return FlyoutMenu;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _flyout_menu = require('../../../../../base/flyout_menu/flyout_menu');
function FlyoutMenu(props) {
    return (0, _jsxruntime.jsx)(_flyout_menu.FlyoutMenu, {
        iconSize: "medium",
        label: props.label,
        trigger: props.trigger ? (easelTriggerProps)=>{
            return props.trigger?.({
                ariaLabel: easelTriggerProps.ariaLabel,
                ariaHasPopup: easelTriggerProps.ariaHasPopup,
                id: easelTriggerProps.id,
                onClick: easelTriggerProps.onRequestToggle,
                pressed: easelTriggerProps.pressed
            });
        } : undefined,
        title: props.title,
        description: props.description,
        ariaLabel: props.ariaLabel,
        children: props.children,
        icon: props.icon,
        iconPosition: props.iconPosition,
        tone: props.tone,
        onOpen: props.onOpen,
        onClose: props.onClose,
        flyoutPlacement: props.flyoutPlacement,
        tooltipLabel: props.tooltipLabel
    });
}
