// @canva/app-ui-kit/dist/cjs/ui/base/scroll/scroll.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get getItemPositions () {
        return _item_positions.getItemPositions;
    },
    get getScrollState () {
        return _scroll_state.getScrollState;
    },
    get isAboveView () {
        return _scroll_into_view.isAboveView;
    },
    get isBelowView () {
        return _scroll_into_view.isBelowView;
    },
    get isFullyInView () {
        return _scroll_into_view.isFullyInView;
    },
    get scrollBy () {
        return _scroll_core.scrollBy;
    },
    get scrollByPage () {
        return _scroll_pages.scrollByPage;
    },
    get scrollIntoView () {
        return _scroll_into_view.scrollIntoView;
    },
    get scrollTo () {
        return _scroll_core.scrollTo;
    },
    get scrollToPage () {
        return _scroll_pages.scrollToPage;
    }
});
const _scroll_state = require("./private/scroll_state");
const _scroll_core = require("./private/scroll_core");
const _scroll_into_view = require("./private/scroll_into_view");
const _scroll_pages = require("./private/scroll_pages");
const _item_positions = require("./private/item_positions");
