// @canva/app-ui-kit/dist/esm/ui/base/placeholder/private/placeholder.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import classNames from 'classnames';
import 'react';
import { useEnableAnimations } from '../../provider/provider';
import { useThemedMedia } from '../../theme/escape_hatch';
import styles, { getStyle } from './placeholder.css';
export const Placeholder = ({ shape, index, disableAnimations })=>{
    const className = classNames(styles.base, {
        [styles.animated]: useEnableAnimations() && disableAnimations !== true
    }, getStyle(shape));
    const style = {};
    if (index != null) style.animationDelay = `calc(${index} * ${styles.flashDelayDuration})`;
    return _jsx("div", {
        className: className,
        style: style
    });
};
export const PlaceholderSvg = ({ shape, index, disableAnimations })=>{
    const opacity = useThemedMedia({
        default: {
            light: '0.07',
            dark: '0.15'
        },
        pulse: {
            light: '0.0025',
            dark: '0.10'
        }
    });
    return _jsxs("svg", {
        className: classNames(styles.svg, {
            [styles.roundCorners]: shape !== 'circle' && shape !== 'sharpRectangle'
        }),
        viewBox: "0 0 100 100",
        preserveAspectRatio: "none",
        opacity: opacity?.default,
        children: [
            shape === 'circle' ? _jsx("circle", {
                cx: "50",
                cy: "50",
                r: "50"
            }) : _jsx("rect", {
                x: "0",
                y: "0",
                width: "100",
                height: "100"
            }),
            useEnableAnimations() && disableAnimations !== true && _jsx("animate", {
                attributeName: "opacity",
                values: `${opacity?.default};${opacity?.pulse};${opacity?.default};${opacity?.default};${opacity?.default}`,
                begin: `${(index ?? 0) * 150}ms`,
                dur: "1400ms",
                repeatCount: "indefinite"
            })
        ]
    });
};
