// @canva/app-ui-kit/dist/cjs/ui/base/scroll/scrollable_items/private/visibility_observer.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get VisibilityRoot () {
        return VisibilityRoot;
    },
    get useVisibilityObserver () {
        return useVisibilityObserver;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const VisibilityContext = _react.createContext({
    observer: undefined,
    callbacks: undefined
});
function useVisibilityObserver({ onVisibilityChange }) {
    const { observer, callbacks } = _react.useContext(VisibilityContext);
    const onVisibilityChangeStable = _react.useEffectEvent((isVisible)=>{
        onVisibilityChange?.(isVisible);
    });
    const ref = _react.useCallback((element)=>{
        callbacks?.set(element, onVisibilityChangeStable);
        observer?.observe(element);
        return ()=>{
            callbacks?.delete(element);
            observer?.unobserve?.(element);
        };
    }, [
        observer,
        callbacks
    ]);
    return {
        ref
    };
}
function VisibilityRoot({ children, disable }) {
    const [observer, setObserver] = _react.useState(undefined);
    const [callbacks, setCallbacks] = _react.useState(undefined);
    const ref = _react.useCallback((element)=>{
        const callbacks = new WeakMap();
        const observer = new IntersectionObserver((entries)=>{
            entries.forEach((entry)=>{
                if (entry.intersectionRatio >= 0.7)
                callbacks?.get(entry.target)?.(true);
                else if (!entry.rootBounds) callbacks?.get(entry.target)?.(false);
                else if (entry.boundingClientRect.width > entry.rootBounds.width)
                callbacks?.get(entry.target)?.(entry.intersectionRatio >= 0.1);
                else callbacks?.get(entry.target)?.(false);
            });
        }, {
            root: element,
            threshold: [
                0,
                0.1,
                0.5,
                0.7,
                1
            ]
        });
        setCallbacks(callbacks);
        setObserver(observer);
        return ()=>{
            observer.disconnect();
        };
    }, []);
    const value = _react.useMemo(()=>({
            observer,
            callbacks
        }), [
        observer,
        callbacks
    ]);
    if (disable) return children({
        ref: noop
    });
    return (0, _jsxruntime.jsx)(VisibilityContext.Provider, {
        value: value,
        children: children({
            ref
        })
    });
}
const noop = ()=>{};
