// @canva/app-ui-kit/dist/cjs/ui/base/tokens/color.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get colorActionBgDisabled () {
        return colorActionBgDisabled;
    },
    get colorActionBorderDisabled () {
        return colorActionBorderDisabled;
    },
    get colorActionCriticalBg () {
        return colorActionCriticalBg;
    },
    get colorActionCriticalBgHovered () {
        return colorActionCriticalBgHovered;
    },
    get colorActionCriticalBgPressed () {
        return colorActionCriticalBgPressed;
    },
    get colorActionCriticalFg () {
        return colorActionCriticalFg;
    },
    get colorActionFgDisabled () {
        return colorActionFgDisabled;
    },
    get colorActionOverlayBg () {
        return colorActionOverlayBg;
    },
    get colorActionOverlayBgDisabled () {
        return colorActionOverlayBgDisabled;
    },
    get colorActionOverlayBgHovered () {
        return colorActionOverlayBgHovered;
    },
    get colorActionOverlayBgPressed () {
        return colorActionOverlayBgPressed;
    },
    get colorActionOverlayFg () {
        return colorActionOverlayFg;
    },
    get colorActionOverlayFgDisabled () {
        return colorActionOverlayFgDisabled;
    },
    get colorActionPrimaryBg () {
        return colorActionPrimaryBg;
    },
    get colorActionPrimaryBgHovered () {
        return colorActionPrimaryBgHovered;
    },
    get colorActionPrimaryBgPressed () {
        return colorActionPrimaryBgPressed;
    },
    get colorActionPrimaryFg () {
        return colorActionPrimaryFg;
    },
    get colorActionSecondaryBg () {
        return colorActionSecondaryBg;
    },
    get colorActionSecondaryBgHovered () {
        return colorActionSecondaryBgHovered;
    },
    get colorActionSecondaryBgPressed () {
        return colorActionSecondaryBgPressed;
    },
    get colorActionSecondaryBorder () {
        return colorActionSecondaryBorder;
    },
    get colorActionSecondaryFg () {
        return colorActionSecondaryFg;
    },
    get colorActionSelectedBg () {
        return colorActionSelectedBg;
    },
    get colorActionSelectedBgHovered () {
        return colorActionSelectedBgHovered;
    },
    get colorActionSelectedBgPressed () {
        return colorActionSelectedBgPressed;
    },
    get colorActionSelectedBorder () {
        return colorActionSelectedBorder;
    },
    get colorActionSelectedFg () {
        return colorActionSelectedFg;
    },
    get colorActionSubtleBgDisabled () {
        return colorActionSubtleBgDisabled;
    },
    get colorActionSubtleFgDisabled () {
        return colorActionSubtleFgDisabled;
    },
    get colorActionTertiaryBg () {
        return colorActionTertiaryBg;
    },
    get colorActionTertiaryBgHovered () {
        return colorActionTertiaryBgHovered;
    },
    get colorActionTertiaryBgPressed () {
        return colorActionTertiaryBgPressed;
    },
    get colorActionTertiaryFg () {
        return colorActionTertiaryFg;
    },
    get colorContentCriticalFg () {
        return colorContentCriticalFg;
    },
    get colorContentFg () {
        return colorContentFg;
    },
    get colorContentInfoFg () {
        return colorContentInfoFg;
    },
    get colorContentPlaceholderFg () {
        return colorContentPlaceholderFg;
    },
    get colorContentPositiveFg () {
        return colorContentPositiveFg;
    },
    get colorContentSubtleFg () {
        return colorContentSubtleFg;
    },
    get colorContentSubtlestFg () {
        return colorContentSubtlestFg;
    },
    get colorContentWarnFg () {
        return colorContentWarnFg;
    },
    get colorControlBg () {
        return colorControlBg;
    },
    get colorControlBgDisabled () {
        return colorControlBgDisabled;
    },
    get colorControlBorder () {
        return colorControlBorder;
    },
    get colorControlBorderDisabled () {
        return colorControlBorderDisabled;
    },
    get colorControlBorderFocused () {
        return colorControlBorderFocused;
    },
    get colorControlBorderHovered () {
        return colorControlBorderHovered;
    },
    get colorControlCriticalBorder () {
        return colorControlCriticalBorder;
    },
    get colorControlSelectedBg () {
        return colorControlSelectedBg;
    },
    get colorControlSelectedBgDisabled () {
        return colorControlSelectedBgDisabled;
    },
    get colorControlSelectedBgHovered () {
        return colorControlSelectedBgHovered;
    },
    get colorControlSelectedBgPressed () {
        return colorControlSelectedBgPressed;
    },
    get colorControlSelectedBorder () {
        return colorControlSelectedBorder;
    },
    get colorControlSelectedFg () {
        return colorControlSelectedFg;
    },
    get colorControlSelectedFgDisabled () {
        return colorControlSelectedFgDisabled;
    },
    get colorControlStrongBorder () {
        return colorControlStrongBorder;
    },
    get colorFeedbackCriticalBg () {
        return colorFeedbackCriticalBg;
    },
    get colorFeedbackCriticalFg () {
        return colorFeedbackCriticalFg;
    },
    get colorFeedbackCriticalSubtleBg () {
        return colorFeedbackCriticalSubtleBg;
    },
    get colorFeedbackCriticalSubtleFg () {
        return colorFeedbackCriticalSubtleFg;
    },
    get colorFeedbackHintBg () {
        return colorFeedbackHintBg;
    },
    get colorFeedbackHintFg () {
        return colorFeedbackHintFg;
    },
    get colorFeedbackHintSubtleBg () {
        return colorFeedbackHintSubtleBg;
    },
    get colorFeedbackHintSubtleFg () {
        return colorFeedbackHintSubtleFg;
    },
    get colorFeedbackInfoBg () {
        return colorFeedbackInfoBg;
    },
    get colorFeedbackInfoFg () {
        return colorFeedbackInfoFg;
    },
    get colorFeedbackInfoSubtleBg () {
        return colorFeedbackInfoSubtleBg;
    },
    get colorFeedbackInfoSubtleFg () {
        return colorFeedbackInfoSubtleFg;
    },
    get colorFeedbackOverlayBg () {
        return colorFeedbackOverlayBg;
    },
    get colorFeedbackOverlayFg () {
        return colorFeedbackOverlayFg;
    },
    get colorFeedbackPositiveBg () {
        return colorFeedbackPositiveBg;
    },
    get colorFeedbackPositiveFg () {
        return colorFeedbackPositiveFg;
    },
    get colorFeedbackPositiveSubtleBg () {
        return colorFeedbackPositiveSubtleBg;
    },
    get colorFeedbackPositiveSubtleFg () {
        return colorFeedbackPositiveSubtleFg;
    },
    get colorFeedbackWarnBg () {
        return colorFeedbackWarnBg;
    },
    get colorFeedbackWarnFg () {
        return colorFeedbackWarnFg;
    },
    get colorFeedbackWarnSubtleBg () {
        return colorFeedbackWarnSubtleBg;
    },
    get colorFeedbackWarnSubtleFg () {
        return colorFeedbackWarnSubtleFg;
    },
    get colorHandleBg () {
        return colorHandleBg;
    },
    get colorHandleBgDisabled () {
        return colorHandleBgDisabled;
    },
    get colorLinkCriticalFg () {
        return colorLinkCriticalFg;
    },
    get colorLinkCriticalFgHovered () {
        return colorLinkCriticalFgHovered;
    },
    get colorLinkCriticalFgPressed () {
        return colorLinkCriticalFgPressed;
    },
    get colorLinkFg () {
        return colorLinkFg;
    },
    get colorLinkFgDisabled () {
        return colorLinkFgDisabled;
    },
    get colorLinkFgHovered () {
        return colorLinkFgHovered;
    },
    get colorLinkFgPressed () {
        return colorLinkFgPressed;
    },
    get colorUiBorder () {
        return colorUiBorder;
    },
    get colorUiBorderFocused () {
        return colorUiBorderFocused;
    },
    get colorUiNeutralBg () {
        return colorUiNeutralBg;
    },
    get colorUiNeutralStrongBg () {
        return colorUiNeutralStrongBg;
    },
    get colorUiNeutralStrongestBg () {
        return colorUiNeutralStrongestBg;
    },
    get colorUiNeutralSubtleBg () {
        return colorUiNeutralSubtleBg;
    },
    get colorUiOverlayBg () {
        return colorUiOverlayBg;
    }
});
const _colorcss = _interop_require_default(require("./private/color.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const colorActionPrimaryBg = _colorcss.default.colorActionPrimaryBg;
const colorActionPrimaryBgHovered = _colorcss.default.colorActionPrimaryBgHovered;
const colorActionPrimaryBgPressed = _colorcss.default.colorActionPrimaryBgPressed;
const colorActionPrimaryFg = _colorcss.default.colorActionPrimaryFg;
const colorActionSecondaryBg = _colorcss.default.colorActionSecondaryBg;
const colorActionSecondaryBgHovered = _colorcss.default.colorActionSecondaryBgHovered;
const colorActionSecondaryBgPressed = _colorcss.default.colorActionSecondaryBgPressed;
const colorActionSecondaryFg = _colorcss.default.colorActionSecondaryFg;
const colorActionSecondaryBorder = _colorcss.default.colorActionSecondaryBorder;
const colorActionTertiaryBg = _colorcss.default.colorActionTertiaryBg;
const colorActionTertiaryBgHovered = _colorcss.default.colorActionTertiaryBgHovered;
const colorActionTertiaryBgPressed = _colorcss.default.colorActionTertiaryBgPressed;
const colorActionTertiaryFg = _colorcss.default.colorActionTertiaryFg;
const colorActionOverlayBg = _colorcss.default.colorActionOverlayBg;
const colorActionOverlayBgHovered = _colorcss.default.colorActionOverlayBgHovered;
const colorActionOverlayBgPressed = _colorcss.default.colorActionOverlayBgPressed;
const colorActionOverlayBgDisabled = _colorcss.default.colorActionOverlayBgDisabled;
const colorActionOverlayFg = _colorcss.default.colorActionOverlayFg;
const colorActionOverlayFgDisabled = _colorcss.default.colorActionOverlayFgDisabled;
const colorActionCriticalBg = _colorcss.default.colorActionCriticalBg;
const colorActionCriticalBgHovered = _colorcss.default.colorActionCriticalBgHovered;
const colorActionCriticalBgPressed = _colorcss.default.colorActionCriticalBgPressed;
const colorActionCriticalFg = _colorcss.default.colorActionCriticalFg;
const colorActionSelectedBg = _colorcss.default.colorActionSelectedBg;
const colorActionSelectedBgHovered = _colorcss.default.colorActionSelectedBgHovered;
const colorActionSelectedBgPressed = _colorcss.default.colorActionSelectedBgPressed;
const colorActionSelectedFg = _colorcss.default.colorActionSelectedFg;
const colorActionSelectedBorder = _colorcss.default.colorActionSelectedBorder;
const colorActionBgDisabled = _colorcss.default.colorActionBgDisabled;
const colorActionFgDisabled = _colorcss.default.colorActionFgDisabled;
const colorActionBorderDisabled = _colorcss.default.colorActionBorderDisabled;
const colorActionSubtleBgDisabled = _colorcss.default.colorActionSubtleBgDisabled;
const colorActionSubtleFgDisabled = _colorcss.default.colorActionSubtleFgDisabled;
const colorFeedbackPositiveBg = _colorcss.default.colorFeedbackPositiveBg;
const colorFeedbackPositiveFg = _colorcss.default.colorFeedbackPositiveFg;
const colorFeedbackPositiveSubtleBg = _colorcss.default.colorFeedbackPositiveSubtleBg;
const colorFeedbackPositiveSubtleFg = _colorcss.default.colorFeedbackPositiveSubtleFg;
const colorFeedbackInfoBg = _colorcss.default.colorFeedbackInfoBg;
const colorFeedbackInfoFg = _colorcss.default.colorFeedbackInfoFg;
const colorFeedbackInfoSubtleBg = _colorcss.default.colorFeedbackInfoSubtleBg;
const colorFeedbackInfoSubtleFg = _colorcss.default.colorFeedbackInfoSubtleFg;
const colorFeedbackWarnBg = _colorcss.default.colorFeedbackWarnBg;
const colorFeedbackWarnFg = _colorcss.default.colorFeedbackWarnFg;
const colorFeedbackWarnSubtleBg = _colorcss.default.colorFeedbackWarnSubtleBg;
const colorFeedbackWarnSubtleFg = _colorcss.default.colorFeedbackWarnSubtleFg;
const colorFeedbackCriticalBg = _colorcss.default.colorFeedbackCriticalBg;
const colorFeedbackCriticalFg = _colorcss.default.colorFeedbackCriticalFg;
const colorFeedbackCriticalSubtleBg = _colorcss.default.colorFeedbackCriticalSubtleBg;
const colorFeedbackCriticalSubtleFg = _colorcss.default.colorFeedbackCriticalSubtleFg;
const colorFeedbackHintBg = _colorcss.default.colorFeedbackHintBg;
const colorFeedbackHintFg = _colorcss.default.colorFeedbackHintFg;
const colorFeedbackHintSubtleBg = _colorcss.default.colorFeedbackHintSubtleBg;
const colorFeedbackHintSubtleFg = _colorcss.default.colorFeedbackHintSubtleFg;
const colorFeedbackOverlayBg = _colorcss.default.colorFeedbackOverlayBg;
const colorFeedbackOverlayFg = _colorcss.default.colorFeedbackOverlayFg;
const colorControlBg = _colorcss.default.colorControlBg;
const colorControlBgDisabled = _colorcss.default.colorControlBgDisabled;
const colorControlBorder = _colorcss.default.colorControlBorder;
const colorControlStrongBorder = _colorcss.default.colorControlStrongBorder;
const colorControlBorderHovered = _colorcss.default.colorControlBorderHovered;
const colorControlBorderFocused = _colorcss.default.colorControlBorderFocused;
const colorControlBorderDisabled = _colorcss.default.colorControlBorderDisabled;
const colorControlCriticalBorder = _colorcss.default.colorControlCriticalBorder;
const colorControlSelectedBg = _colorcss.default.colorControlSelectedBg;
const colorControlSelectedBgHovered = _colorcss.default.colorControlSelectedBgHovered;
const colorControlSelectedBgPressed = _colorcss.default.colorControlSelectedBgPressed;
const colorControlSelectedBgDisabled = _colorcss.default.colorControlSelectedBgDisabled;
const colorControlSelectedFg = _colorcss.default.colorControlSelectedFg;
const colorControlSelectedFgDisabled = _colorcss.default.colorControlSelectedFgDisabled;
const colorControlSelectedBorder = _colorcss.default.colorControlSelectedBorder;
const colorHandleBg = _colorcss.default.colorHandleBg;
const colorHandleBgDisabled = _colorcss.default.colorHandleBgDisabled;
const colorLinkFg = _colorcss.default.colorLinkFg;
const colorLinkFgHovered = _colorcss.default.colorLinkFgHovered;
const colorLinkFgPressed = _colorcss.default.colorLinkFgPressed;
const colorLinkFgDisabled = _colorcss.default.colorLinkFgDisabled;
const colorLinkCriticalFg = _colorcss.default.colorLinkCriticalFg;
const colorLinkCriticalFgHovered = _colorcss.default.colorLinkCriticalFgHovered;
const colorLinkCriticalFgPressed = _colorcss.default.colorLinkCriticalFgPressed;
const colorContentFg = _colorcss.default.colorContentFg;
const colorContentSubtleFg = _colorcss.default.colorContentSubtleFg;
const colorContentSubtlestFg = _colorcss.default.colorContentSubtlestFg;
const colorContentPlaceholderFg = _colorcss.default.colorContentPlaceholderFg;
const colorContentPositiveFg = _colorcss.default.colorContentPositiveFg;
const colorContentCriticalFg = _colorcss.default.colorContentCriticalFg;
const colorContentInfoFg = _colorcss.default.colorContentInfoFg;
const colorContentWarnFg = _colorcss.default.colorContentWarnFg;
const colorUiBorder = _colorcss.default.colorUiBorder;
const colorUiBorderFocused = _colorcss.default.colorUiBorderFocused;
const colorUiNeutralBg = _colorcss.default.colorUiNeutralBg;
const colorUiNeutralSubtleBg = _colorcss.default.colorUiNeutralSubtleBg;
const colorUiNeutralStrongBg = _colorcss.default.colorUiNeutralStrongBg;
const colorUiNeutralStrongestBg = _colorcss.default.colorUiNeutralStrongestBg;
const colorUiOverlayBg = _colorcss.default.colorUiOverlayBg;
