// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/internal/handle.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Handle", {
    enumerable: true,
    get: function() {
        return Handle;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../box/box');
const _use_isomorphic_layout_effect = require('../../hydration/use_isomorphic_layout_effect');
const _pin = require('../../surface/pin/pin');
const _swatch = require('../../swatch/swatch');
const _handlecss = _interop_require_default(require("./handle.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function Handle({ color, active = false, focused = false, preview = false, className, style }) {
    const pinRef = _react.useRef(null);
    (0, _use_isomorphic_layout_effect.useIsomorphicLayoutEffect)(()=>{
        preview && pinRef.current?.update();
    });
    const shouldScale = focused || active;
    const handleClassName = (0, _classnames.default)(className, _handlecss.default.handle, {
        [_handlecss.default.handleScaled]: shouldScale
    });
    const swatchClassName = (0, _classnames.default)(_handlecss.default.handleSwatch, {
        [_handlecss.default.handleSwatchScaled]: shouldScale
    });
    return (0, _jsxruntime.jsx)(_pin.Pin, {
        ref: pinRef,
        open: preview,
        placement: "top-center",
        reference: (0, _jsxruntime.jsx)("div", {
            className: handleClassName,
            style: style,
            children: (0, _jsxruntime.jsx)("div", {
                className: swatchClassName,
                style: {
                    backgroundColor: color
                }
            })
        }),
        children: (0, _jsxruntime.jsx)(_box.Box, {
            className: _handlecss.default.tooltip,
            border: "low",
            borderRadius: "elementRound",
            background: "elevationSurfaceFloating",
            shadow: "elevationSurfaceFloating",
            position: "absolute",
            padding: "1u",
            children: (0, _jsxruntime.jsx)(_swatch.Swatch, {
                fill: [
                    color
                ],
                size: "xsmall"
            })
        })
    });
}
