// @canva/app-ui-kit/dist/esm/ui/base/tokens/color.js
import color from './private/color.css';
export const colorActionPrimaryBg = color.colorActionPrimaryBg;
export const colorActionPrimaryBgHovered = color.colorActionPrimaryBgHovered;
export const colorActionPrimaryBgPressed = color.colorActionPrimaryBgPressed;
export const colorActionPrimaryFg = color.colorActionPrimaryFg;
export const colorActionSecondaryBg = color.colorActionSecondaryBg;
export const colorActionSecondaryBgHovered = color.colorActionSecondaryBgHovered;
export const colorActionSecondaryBgPressed = color.colorActionSecondaryBgPressed;
export const colorActionSecondaryFg = color.colorActionSecondaryFg;
export const colorActionSecondaryBorder = color.colorActionSecondaryBorder;
export const colorActionTertiaryBg = color.colorActionTertiaryBg;
export const colorActionTertiaryBgHovered = color.colorActionTertiaryBgHovered;
export const colorActionTertiaryBgPressed = color.colorActionTertiaryBgPressed;
export const colorActionTertiaryFg = color.colorActionTertiaryFg;
export const colorActionOverlayBg = color.colorActionOverlayBg;
export const colorActionOverlayBgHovered = color.colorActionOverlayBgHovered;
export const colorActionOverlayBgPressed = color.colorActionOverlayBgPressed;
export const colorActionOverlayBgDisabled = color.colorActionOverlayBgDisabled;
export const colorActionOverlayFg = color.colorActionOverlayFg;
export const colorActionOverlayFgDisabled = color.colorActionOverlayFgDisabled;
export const colorActionCriticalBg = color.colorActionCriticalBg;
export const colorActionCriticalBgHovered = color.colorActionCriticalBgHovered;
export const colorActionCriticalBgPressed = color.colorActionCriticalBgPressed;
export const colorActionCriticalFg = color.colorActionCriticalFg;
export const colorActionSelectedBg = color.colorActionSelectedBg;
export const colorActionSelectedBgHovered = color.colorActionSelectedBgHovered;
export const colorActionSelectedBgPressed = color.colorActionSelectedBgPressed;
export const colorActionSelectedFg = color.colorActionSelectedFg;
export const colorActionSelectedBorder = color.colorActionSelectedBorder;
export const colorActionBgDisabled = color.colorActionBgDisabled;
export const colorActionFgDisabled = color.colorActionFgDisabled;
export const colorActionBorderDisabled = color.colorActionBorderDisabled;
export const colorActionSubtleBgDisabled = color.colorActionSubtleBgDisabled;
export const colorActionSubtleFgDisabled = color.colorActionSubtleFgDisabled;
export const colorFeedbackPositiveBg = color.colorFeedbackPositiveBg;
export const colorFeedbackPositiveFg = color.colorFeedbackPositiveFg;
export const colorFeedbackPositiveSubtleBg = color.colorFeedbackPositiveSubtleBg;
export const colorFeedbackPositiveSubtleFg = color.colorFeedbackPositiveSubtleFg;
export const colorFeedbackInfoBg = color.colorFeedbackInfoBg;
export const colorFeedbackInfoFg = color.colorFeedbackInfoFg;
export const colorFeedbackInfoSubtleBg = color.colorFeedbackInfoSubtleBg;
export const colorFeedbackInfoSubtleFg = color.colorFeedbackInfoSubtleFg;
export const colorFeedbackWarnBg = color.colorFeedbackWarnBg;
export const colorFeedbackWarnFg = color.colorFeedbackWarnFg;
export const colorFeedbackWarnSubtleBg = color.colorFeedbackWarnSubtleBg;
export const colorFeedbackWarnSubtleFg = color.colorFeedbackWarnSubtleFg;
export const colorFeedbackCriticalBg = color.colorFeedbackCriticalBg;
export const colorFeedbackCriticalFg = color.colorFeedbackCriticalFg;
export const colorFeedbackCriticalSubtleBg = color.colorFeedbackCriticalSubtleBg;
export const colorFeedbackCriticalSubtleFg = color.colorFeedbackCriticalSubtleFg;
export const colorFeedbackHintBg = color.colorFeedbackHintBg;
export const colorFeedbackHintFg = color.colorFeedbackHintFg;
export const colorFeedbackHintSubtleBg = color.colorFeedbackHintSubtleBg;
export const colorFeedbackHintSubtleFg = color.colorFeedbackHintSubtleFg;
export const colorFeedbackOverlayBg = color.colorFeedbackOverlayBg;
export const colorFeedbackOverlayFg = color.colorFeedbackOverlayFg;
export const colorControlBg = color.colorControlBg;
export const colorControlBgDisabled = color.colorControlBgDisabled;
export const colorControlBorder = color.colorControlBorder;
export const colorControlStrongBorder = color.colorControlStrongBorder;
export const colorControlBorderHovered = color.colorControlBorderHovered;
export const colorControlBorderFocused = color.colorControlBorderFocused;
export const colorControlBorderDisabled = color.colorControlBorderDisabled;
export const colorControlCriticalBorder = color.colorControlCriticalBorder;
export const colorControlSelectedBg = color.colorControlSelectedBg;
export const colorControlSelectedBgHovered = color.colorControlSelectedBgHovered;
export const colorControlSelectedBgPressed = color.colorControlSelectedBgPressed;
export const colorControlSelectedBgDisabled = color.colorControlSelectedBgDisabled;
export const colorControlSelectedFg = color.colorControlSelectedFg;
export const colorControlSelectedFgDisabled = color.colorControlSelectedFgDisabled;
export const colorControlSelectedBorder = color.colorControlSelectedBorder;
export const colorHandleBg = color.colorHandleBg;
export const colorHandleBgDisabled = color.colorHandleBgDisabled;
export const colorLinkFg = color.colorLinkFg;
export const colorLinkFgHovered = color.colorLinkFgHovered;
export const colorLinkFgPressed = color.colorLinkFgPressed;
export const colorLinkFgDisabled = color.colorLinkFgDisabled;
export const colorLinkCriticalFg = color.colorLinkCriticalFg;
export const colorLinkCriticalFgHovered = color.colorLinkCriticalFgHovered;
export const colorLinkCriticalFgPressed = color.colorLinkCriticalFgPressed;
export const colorContentFg = color.colorContentFg;
export const colorContentSubtleFg = color.colorContentSubtleFg;
export const colorContentSubtlestFg = color.colorContentSubtlestFg;
export const colorContentPlaceholderFg = color.colorContentPlaceholderFg;
export const colorContentPositiveFg = color.colorContentPositiveFg;
export const colorContentCriticalFg = color.colorContentCriticalFg;
export const colorContentInfoFg = color.colorContentInfoFg;
export const colorContentWarnFg = color.colorContentWarnFg;
export const colorUiBorder = color.colorUiBorder;
export const colorUiBorderFocused = color.colorUiBorderFocused;
export const colorUiNeutralBg = color.colorUiNeutralBg;
export const colorUiNeutralSubtleBg = color.colorUiNeutralSubtleBg;
export const colorUiNeutralStrongBg = color.colorUiNeutralStrongBg;
export const colorUiNeutralStrongestBg = color.colorUiNeutralStrongestBg;
export const colorUiOverlayBg = color.colorUiOverlayBg;
