// @canva/app-ui-kit/dist/esm/ui/base/tokens/primitive/color.js
import primitiveColor from './private/color.css';
export const colorTeal01 = primitiveColor.colorTeal01;
export const colorTeal02 = primitiveColor.colorTeal02;
export const colorTeal03 = primitiveColor.colorTeal03;
export const colorTeal04 = primitiveColor.colorTeal04;
export const colorGreen01 = primitiveColor.colorGreen01;
export const colorGreen02 = primitiveColor.colorGreen02;
export const colorGreen03 = primitiveColor.colorGreen03;
export const colorGreen04 = primitiveColor.colorGreen04;
export const colorGreen05 = primitiveColor.colorGreen05;
export const colorGreen06 = primitiveColor.colorGreen06;
export const colorGreen07 = primitiveColor.colorGreen07;
export const colorGreen08 = primitiveColor.colorGreen08;
export const colorGreen09 = primitiveColor.colorGreen09;
export const colorGreen10 = primitiveColor.colorGreen10;
export const colorPurple01 = primitiveColor.colorPurple01;
export const colorPurple02 = primitiveColor.colorPurple02;
export const colorPurple03 = primitiveColor.colorPurple03;
export const colorPurple04 = primitiveColor.colorPurple04;
export const colorPurple05 = primitiveColor.colorPurple05;
export const colorPurple06 = primitiveColor.colorPurple06;
export const colorPurple07 = primitiveColor.colorPurple07;
export const colorPurple08 = primitiveColor.colorPurple08;
export const colorPurple09 = primitiveColor.colorPurple09;
export const colorPurple10 = primitiveColor.colorPurple10;
export const colorPurpleA00 = primitiveColor.colorPurpleA00;
export const colorPurpleA01 = primitiveColor.colorPurpleA01;
export const colorPurpleA02 = primitiveColor.colorPurpleA02;
export const colorPurpleA03 = primitiveColor.colorPurpleA03;
export const colorPurpleA04 = primitiveColor.colorPurpleA04;
export const colorPurpleA05 = primitiveColor.colorPurpleA05;
export const colorPurpleA06 = primitiveColor.colorPurpleA06;
export const colorPurpleA07 = primitiveColor.colorPurpleA07;
export const colorPurpleA08 = primitiveColor.colorPurpleA08;
export const colorPurpleA09 = primitiveColor.colorPurpleA09;
export const colorPurpleA10 = primitiveColor.colorPurpleA10;
export const colorRed01 = primitiveColor.colorRed01;
export const colorRed02 = primitiveColor.colorRed02;
export const colorRed03 = primitiveColor.colorRed03;
export const colorRed04 = primitiveColor.colorRed04;
export const colorRed05 = primitiveColor.colorRed05;
export const colorRed06 = primitiveColor.colorRed06;
export const colorRed07 = primitiveColor.colorRed07;
export const colorRed08 = primitiveColor.colorRed08;
export const colorRed09 = primitiveColor.colorRed09;
export const colorRed10 = primitiveColor.colorRed10;
export const colorBlue01 = primitiveColor.colorBlue01;
export const colorBlue02 = primitiveColor.colorBlue02;
export const colorBlue03 = primitiveColor.colorBlue03;
export const colorBlue04 = primitiveColor.colorBlue04;
export const colorBlue05 = primitiveColor.colorBlue05;
export const colorBlue06 = primitiveColor.colorBlue06;
export const colorBlue07 = primitiveColor.colorBlue07;
export const colorBlue08 = primitiveColor.colorBlue08;
export const colorBlue09 = primitiveColor.colorBlue09;
export const colorBlue10 = primitiveColor.colorBlue10;
export const colorYellow01 = primitiveColor.colorYellow01;
export const colorYellow02 = primitiveColor.colorYellow02;
export const colorYellow03 = primitiveColor.colorYellow03;
export const colorYellow04 = primitiveColor.colorYellow04;
export const colorYellow05 = primitiveColor.colorYellow05;
export const colorYellow06 = primitiveColor.colorYellow06;
export const colorYellow07 = primitiveColor.colorYellow07;
export const colorYellow08 = primitiveColor.colorYellow08;
export const colorYellow09 = primitiveColor.colorYellow09;
export const colorYellow10 = primitiveColor.colorYellow10;
export const colorWhiteA00 = primitiveColor.colorWhiteA00;
export const colorWhiteA01 = primitiveColor.colorWhiteA01;
export const colorWhiteA02 = primitiveColor.colorWhiteA02;
export const colorWhiteA03 = primitiveColor.colorWhiteA03;
export const colorWhiteA04 = primitiveColor.colorWhiteA04;
export const colorWhiteA05 = primitiveColor.colorWhiteA05;
export const colorWhiteA06 = primitiveColor.colorWhiteA06;
export const colorWhiteA07 = primitiveColor.colorWhiteA07;
export const colorWhiteA08 = primitiveColor.colorWhiteA08;
export const colorWhiteA09 = primitiveColor.colorWhiteA09;
export const colorWhiteA10 = primitiveColor.colorWhiteA10;
export const colorGrey01 = primitiveColor.colorGrey01;
export const colorGrey02 = primitiveColor.colorGrey02;
export const colorGrey03 = primitiveColor.colorGrey03;
export const colorGrey04 = primitiveColor.colorGrey04;
export const colorGrey05 = primitiveColor.colorGrey05;
export const colorGrey06 = primitiveColor.colorGrey06;
export const colorGrey07 = primitiveColor.colorGrey07;
export const colorGrey08 = primitiveColor.colorGrey08;
export const colorGrey09 = primitiveColor.colorGrey09;
export const colorGrey10 = primitiveColor.colorGrey10;
export const colorBlackA00 = primitiveColor.colorBlackA00;
export const colorBlackA01 = primitiveColor.colorBlackA01;
export const colorBlackA02 = primitiveColor.colorBlackA02;
export const colorBlackA03 = primitiveColor.colorBlackA03;
export const colorBlackA04 = primitiveColor.colorBlackA04;
export const colorBlackA05 = primitiveColor.colorBlackA05;
export const colorBlackA06 = primitiveColor.colorBlackA06;
export const colorBlackA07 = primitiveColor.colorBlackA07;
export const colorBlackA08 = primitiveColor.colorBlackA08;
export const colorBlackA09 = primitiveColor.colorBlackA09;
export const colorBlackA10 = primitiveColor.colorBlackA10;
