diff --git a/node_modules/jwks-rsa/src/integrations/passport.js b/node_modules/jwks-rsa/src/integrations/passport.js
index 79d75da..2d27d11 100644
--- a/node_modules/jwks-rsa/src/integrations/passport.js
+++ b/node_modules/jwks-rsa/src/integrations/passport.js
@@ -1,4 +1,4 @@
-const jose = require('jose');
+const jose = {};
 const { ArgumentError } = require('../errors');
 const { JwksClient } = require('../JwksClient');
 const supportedAlg = require('./config');
diff --git a/node_modules/jwks-rsa/src/utils.js b/node_modules/jwks-rsa/src/utils.js
index 3b82d11..fdc4a57 100644
--- a/node_modules/jwks-rsa/src/utils.js
+++ b/node_modules/jwks-rsa/src/utils.js
@@ -1,4 +1,4 @@
-const jose = require('jose');
+const jose = {};
 const JwksError = require('./errors/JwksError');
 
 function resolveAlg(jwk) {
