// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/flyout/flyout.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Flyout", {
    enumerable: true,
    get: function() {
        return Flyout;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _flyout = require('../../../../../base/surface/flyout/flyout');
function Flyout(props) {
    const headerStart = typeof props.headerStart === 'function' ? props.headerStart() : props.headerStart;
    const headerEnd = typeof props.headerEnd === 'function' ? props.headerEnd() : props.headerEnd;
    return (0, _jsxruntime.jsx)(_flyout.Flyout, {
        title: props.title,
        description: props.description,
        headerStart: headerStart,
        headerEnd: headerEnd,
        open: props.open,
        onRequestClose: props.onRequestClose,
        onCloseComplete: props.onCloseComplete,
        placement: props.placement,
        width: props.width ?? '32u',
        trigger: props.trigger,
        footer: props.footer,
        children: props.children
    });
}
