// @canva/app-ui-kit/dist/esm/ui/base/color_picker/internal/color_picker_provider.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { HsvColor } from '../hsv_color/hsv_color';
const ColorPickerHueContext = React.createContext(undefined);
const ColorPickerSaturationContext = React.createContext(undefined);
const ColorPickerValueContext = React.createContext(undefined);
const ColorPickerAlphaContext = React.createContext(undefined);
const ColorPickerControlsContext = React.createContext(undefined);
export function ColorPickerHueProvider({ hue, children }) {
    return _jsx(ColorPickerHueContext.Provider, {
        value: hue,
        children: children
    });
}
export function ColorPickerSaturationProvider({ saturation, children }) {
    return _jsx(ColorPickerSaturationContext.Provider, {
        value: saturation,
        children: children
    });
}
export function ColorPickerValueProvider({ value, children }) {
    return _jsx(ColorPickerValueContext.Provider, {
        value: value,
        children: children
    });
}
export function ColorPickerAlphaProvider({ alpha, children }) {
    return _jsx(ColorPickerAlphaContext.Provider, {
        value: alpha ?? 1,
        children: children
    });
}
export function ColorPickerControlsProvider({ setColor, onChangeStart, onChangeComplete, children }) {
    const value = React.useMemo(()=>({
            setColor,
            onChangeStart,
            onChangeComplete
        }), [
        setColor,
        onChangeStart,
        onChangeComplete
    ]);
    return _jsx(ColorPickerControlsContext.Provider, {
        value: value,
        children: children
    });
}
export function useColorPickerHue() {
    return React.useContext(ColorPickerHueContext);
}
export function useColorPickerSaturation() {
    return React.useContext(ColorPickerSaturationContext);
}
export function useColorPickerValue() {
    return React.useContext(ColorPickerValueContext);
}
export function useColorPickerAlpha() {
    return React.useContext(ColorPickerAlphaContext);
}
export function useColorPickerColor() {
    const h = useColorPickerHue();
    const s = useColorPickerSaturation();
    const v = useColorPickerValue();
    const a = useColorPickerAlpha();
    if (h == null || s == null || v == null) return undefined;
    return new HsvColor(h, s, v, a);
}
export function useColorPickerHex() {
    const color = useColorPickerColor();
    return color?.toRgb().toHexString();
}
export function useColorPickerControls() {
    return React.useContext(ColorPickerControlsContext);
}
