// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/private/color_picker_flyout.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get ColorPickerFlyout () {
        return ColorPickerFlyout;
    },
    get StatelessColorPickerFlyout () {
        return StatelessColorPickerFlyout;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../box/box');
const _flyout = require('../../surface/flyout/flyout');
const _swatch = require('../../swatch/swatch');
const _color_picker = require("./color_picker");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ColorPickerFlyout = ({ trigger, color, alpha, ariaLabel, ariaLabelledBy, onChange, onChangeStart, onChangeComplete, onStartEyedropper, onOpen, onClose, onDeleteColor, header, body, footer })=>{
    const [open, setOpen] = _react.useState(false);
    const toggleFlyout = _react.useCallback(()=>{
        if (open)
            onClose?.();
        else
            onOpen?.();
        setOpen((prev)=>!prev);
    }, [
        open,
        onOpen,
        onClose,
        setOpen
    ]);
    const closeFlyout = _react.useCallback(()=>{
        onClose?.();
        setOpen(false);
    }, [
        onClose,
        setOpen
    ]);
    const triggerId = _react.useId();
    return (0, _jsxruntime.jsx)(StatelessColorPickerFlyout, {
        trigger: trigger,
        open: open,
        id: triggerId,
        onClick: toggleFlyout,
        onRequestClose: closeFlyout,
        ariaLabel: ariaLabel,
        ariaLabelledBy: ariaLabelledBy,
        color: color,
        alpha: alpha,
        onChange: onChange,
        onChangeStart: onChangeStart,
        onChangeComplete: onChangeComplete,
        onStartEyedropper: onStartEyedropper,
        onDeleteColor: onDeleteColor,
        header: header,
        body: body,
        footer: footer
    });
};
function StatelessColorPickerFlyout(props) {
    const { trigger = (props)=> (0, _jsxruntime.jsx)(_swatch.AddColorButton, {
            size: "small",
            ...props
        }), open, id, ariaLabel, ariaLabelledBy, color, alpha, onClick, onRequestClose, onChange, onChangeStart, onChangeComplete, onStartEyedropper, onDeleteColor, header, body, footer } = props;
    return (0, _jsxruntime.jsx)(_flyout.Flyout, {
        role: "dialog",
        ariaLabelledBy: id,
        placement: "bottom-start",
        open: open,
        onRequestClose: onRequestClose,
        captureFocus: true,
        width: "40u",
        trigger: (triggerProps)=>trigger({
                id,
                ariaLabel,
                ariaLabelledBy,
                onClick,
                ...triggerProps
            }),
        children: (0, _jsxruntime.jsx)(_box.Box, {
            padding: "2u",
            children: (0, _jsxruntime.jsx)(_color_picker.ColorPicker, {
                color: color,
                alpha: alpha,
                onChange: onChange,
                onChangeStart: onChangeStart,
                onChangeComplete: onChangeComplete,
                onStartEyedropper: onStartEyedropper,
                onDeleteColor: onDeleteColor,
                header: header,
                body: body,
                footer: footer
            })
        })
    });
}
