// @canva/app-ui-kit/dist/esm/ui/base/scroll/scrollable_items/private/visibility_observer.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
const VisibilityContext = React.createContext({
    observer: undefined,
    callbacks: undefined
});
export function useVisibilityObserver({ onVisibilityChange }) {
    const { observer, callbacks } = React.useContext(VisibilityContext);
    const onVisibilityChangeStable = React.useEffectEvent((isVisible)=>{
        onVisibilityChange?.(isVisible);
    });
    const ref = React.useCallback((element)=>{
        callbacks?.set(element, onVisibilityChangeStable);
        observer?.observe(element);
        return ()=>{
            callbacks?.delete(element);
            observer?.unobserve?.(element);
        };
    }, [
        observer,
        callbacks
    ]);
    return {
        ref
    };
}
export function VisibilityRoot({ children, disable }) {
    const [observer, setObserver] = React.useState(undefined);
    const [callbacks, setCallbacks] = React.useState(undefined);
    const ref = React.useCallback((element)=>{
        const callbacks = new WeakMap();
        const observer = new IntersectionObserver((entries)=>{
            entries.forEach((entry)=>{
                if (entry.intersectionRatio >= 0.7)
                callbacks?.get(entry.target)?.(true);
                else if (!entry.rootBounds) callbacks?.get(entry.target)?.(false);
                else if (entry.boundingClientRect.width > entry.rootBounds.width)
                callbacks?.get(entry.target)?.(entry.intersectionRatio >= 0.1);
                else callbacks?.get(entry.target)?.(false);
            });
        }, {
            root: element,
            threshold: [
                0,
                0.1,
                0.5,
                0.7,
                1
            ]
        });
        setCallbacks(callbacks);
        setObserver(observer);
        return ()=>{
            observer.disconnect();
        };
    }, []);
    const value = React.useMemo(()=>({
            observer,
            callbacks
        }), [
        observer,
        callbacks
    ]);
    if (disable) return children({
        ref: noop
    });
    return _jsx(VisibilityContext.Provider, {
        value: value,
        children: children({
            ref
        })
    });
}
const noop = ()=>{};
