// @canva/app-ui-kit/dist/cjs/ui/base/focus/focus.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get ApplyFocusHistory () {
        return ApplyFocusHistory;
    },
    get AutoFocus () {
        return AutoFocus;
    },
    get FocusManagerProvider () {
        return _focus_manager.FocusManagerProvider;
    },
    get FocusScope () {
        return FocusScope;
    },
    get FocusWhen () {
        return FocusWhen;
    },
    get ReturnFocus () {
        return ReturnFocus;
    },
    get SyncFocus () {
        return SyncFocus;
    },
    get WithFocusManager () {
        return _focus_manager.WithFocusManager;
    },
    get escapeFocusRecovery () {
        return escapeFocusRecovery;
    },
    get findFocusable () {
        return _focus_utils.findFocusable;
    },
    get isFocusable () {
        return _focus_utils.isFocusable;
    },
    get useAutoFocus () {
        return useAutoFocus;
    },
    get useFocusManager () {
        return _focus_manager.useFocusManager;
    },
    get useFocusScope () {
        return useFocusScope;
    },
    get useFocusVisible () {
        return useFocusVisible;
    },
    get useFocusWhen () {
        return useFocusWhen;
    },
    get useReturnFocus () {
        return useReturnFocus;
    },
    get useSyncFocus () {
        return useSyncFocus;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _compose_refs = require('../../../base/react/compose_refs');
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../a11y/access_mode_state/access_mode_state');
const _use_isomorphic_layout_effect = require('../hydration/use_isomorphic_layout_effect');
const _focuscss = _interop_require_default(require("./focus.css"));
const _focus_history = require("./focus_history");
const _focus_manager = require("./focus_manager");
const _focus_utils = require("./internal/focus_utils");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function ApplyFocusHistory() {
    _react.useEffect(()=>(0, _focus_history.registerEvents)(document), []);
    return null;
}
function escapeFocusRecovery(element) {
    element.focus();
    element.blur();
}
function useFocusVisible() {
    return !(0, _access_mode_state.useIsMouseAccessMode)();
}
function useFocusScope({ onEnter, onLeave } = {}) {
    const ref = _react.useRef(null);
    const isFocusWithinRef = _react.useRef(false);
    const onFocusCapture = _react.useCallback((e)=>{
        if (_focus_history.FocusHistory.enabled && e.target instanceof HTMLElement) _focus_history.FocusHistory.replace((0, _focus_history.createFocusState)(e.target, ref.current));
        if (isFocusWithinRef.current) return;
        isFocusWithinRef.current = true;
        onEnter?.(e);
    }, [
        onEnter
    ]);
    const onBlur = _react.useCallback((e)=>{
        if (!(e.relatedTarget instanceof HTMLElement)) return;
        if (ref.current?.contains(e.relatedTarget)) return;
        isFocusWithinRef.current = false;
        onLeave?.(e);
    }, [
        onLeave
    ]);
    const onLeaveRef = _react.useRef(onLeave);
    (0, _use_isomorphic_layout_effect.useIsomorphicLayoutEffect)(()=>{
        onLeaveRef.current = onLeave;
    }, [
        onLeave
    ]);
    _react.useEffect(()=>{
        return ()=>{
            if (isFocusWithinRef.current)
                onLeaveRef.current?.();
        };
    }, []);
    return {
        ref,
        onFocusCapture,
        onBlur
    };
}
function FocusScope(props) {
    const { onFocus: propOnFocus, onBlur: propOnBlur, ref } = props;
    const { ref: scopeRef, onBlur: scopeOnBlur, onFocusCapture: scopeOnFocusCapture } = useFocusScope(props);
    const { ref: focusManagerRef, onFocusCapture: managerOnFocusCapture, onBlur: managerOnBlur, focusManager } = (0, _focus_manager.useCreateFocusManager)();
    const onFocusCapture = _react.useCallback((e)=>{
        managerOnFocusCapture(e);
        scopeOnFocusCapture(e);
    }, [
        managerOnFocusCapture,
        scopeOnFocusCapture
    ]);
    const onBlur = _react.useCallback((e)=>{
        managerOnBlur(e);
        scopeOnBlur(e);
        propOnBlur?.(e);
    }, [
        managerOnBlur,
        scopeOnBlur,
        propOnBlur
    ]);
    _react.useImperativeHandle(ref, ()=>focusManager, [
        focusManager
    ]);
    return (0, _jsxruntime.jsx)(_focus_manager.FocusManagerProvider, {
        value: focusManager,
        children: (0, _jsxruntime.jsx)("div", {
            ref: (0, _compose_refs.composeRefs)(scopeRef, focusManagerRef),
            id: props.id,
            className: _focuscss.default.nolayout,
            onFocusCapture: onFocusCapture,
            onFocus: propOnFocus,
            onBlur: onBlur,
            children: props.children
        })
    });
}
function getFocusableFrom(el, focusManager) {
    if (el instanceof HTMLElement) return (0, _focus_utils.isFocusable)(el) ? el : (0, _focus_utils.findFocusable)(el);
    return focusManager?.focusables[0];
}
function useAutoFocus() {
    const wrapperRef = _react.useRef(null);
    const focusManager = (0, _focus_manager.useFocusManager)();
    useFocusWhen(true, ()=>getFocusableFrom(wrapperRef.current, focusManager));
    return {
        ref: wrapperRef
    };
}
function AutoFocus(props) {
    const { ref } = useAutoFocus();
    return props.children && (0, _jsxruntime.jsx)("div", {
        ref: ref,
        className: _focuscss.default.nolayout,
        children: props.children
    });
}
function useReturnFocus(props) {
    const focusManager = (0, _focus_manager.useFocusManager)();
    const { to } = props;
    const toRef = _react.useRef(to);
    (0, _use_isomorphic_layout_effect.useIsomorphicLayoutEffect)(()=>{
        toRef.current = to;
    }, [
        to
    ]);
    _react.useEffect(()=>{
        return ()=>{
            if (focusManager?.focused == null)
                return;
            const element = typeof toRef.current === 'string' ? document.getElementById(toRef.current) : toRef.current();
            if (!element) return;
            element.focus();
        };
    }, [
        focusManager
    ]);
}
function ReturnFocus(props) {
    useReturnFocus(props);
    return null;
}
function useFocusWhen(condition, getFocusable) {
    const lastValueRef = _react.useRef(undefined);
    _react.useEffect(()=>{
        if (condition === lastValueRef.current) return;
        if (condition === true) getFocusable()?.focus();
        lastValueRef.current = condition;
    }, [
        condition,
        getFocusable
    ]);
}
function FocusWhen(props) {
    const wrapperRef = _react.useRef(null);
    const focusManager = (0, _focus_manager.useFocusManager)();
    useFocusWhen(props.condition, ()=>getFocusableFrom(wrapperRef.current, focusManager));
    return (0, _jsxruntime.jsx)("div", {
        ref: wrapperRef,
        className: _focuscss.default.nolayout,
        children: props.children
    });
}
function useSyncFocus(index, onFocusIndex) {
    const wrapperRef = _react.useRef(null);
    const focusManager = (0, _focus_manager.useFocusManager)();
    const lastIndexRef = _react.useRef(index);
    _react.useEffect(()=>{
        if (index === lastIndexRef.current) return;
        lastIndexRef.current = index;
        if (wrapperRef.current instanceof HTMLElement) return (0, _focus_utils.findFocusables)(wrapperRef.current)[index]?.focus();
        if (focusManager == null) return;
        focusManager.focusAt(index);
    }, [
        index,
        focusManager
    ]);
    const onFocusIndexRef = _react.useRef(onFocusIndex);
    (0, _use_isomorphic_layout_effect.useIsomorphicLayoutEffect)(()=>{
        onFocusIndexRef.current = onFocusIndex;
    }, [
        onFocusIndex
    ]);
    const onFocus = _react.useCallback((e)=>{
        const focusables = wrapperRef.current instanceof HTMLElement ? (0, _focus_utils.findFocusables)(wrapperRef.current) : focusManager?.focusables;
        const index = focusables?.findIndex((el)=>el === e.target) ?? -1;
        onFocusIndexRef.current?.(index);
    }, [
        focusManager
    ]);
    return {
        ref: wrapperRef,
        onFocus
    };
}
function SyncFocus(props) {
    const { ref, onFocus } = useSyncFocus(props.index, props.onFocusIndex);
    return (0, _jsxruntime.jsx)("div", {
        ref: ref,
        onFocus: onFocus,
        className: _focuscss.default.nolayout,
        children: props.children
    });
}
