// @canva/app-ui-kit/dist/cjs/ui/base/focus/focus_manager.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get FocusManagerProvider () {
        return FocusManagerProvider;
    },
    get WithFocusManager () {
        return WithFocusManager;
    },
    get useCreateFocusManager () {
        return useCreateFocusManager;
    },
    get useFocusManager () {
        return useFocusManager;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _compose_refs = require('../../../base/react/compose_refs');
const _react = _interop_require_wildcard(require("react"));
const _focuscss = _interop_require_default(require("./focus.css"));
const _focus_utils = require("./internal/focus_utils");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function useMutationObserver(callback) {
    const ref = _react.useRef(null);
    const observerRef = _react.useRef(undefined);
    _react.useEffect(()=>{
        if (!ref.current) return;
        observerRef.current = new MutationObserver(callback);
        observerRef.current.observe(ref.current, {
            childList: true,
            subtree: true
        });
        callback([], observerRef.current);
        return ()=>observerRef.current?.disconnect();
    }, [
        callback
    ]);
    return {
        ref
    };
}
function useCreateFocusManager() {
    const rootRef = _react.useRef(null);
    const focusablesRef = _react.useRef([]);
    const focusedRef = _react.useRef(null);
    const onMutation = _react.useCallback(()=>{
        if (rootRef.current) focusablesRef.current = (0, _focus_utils.findFocusables)(rootRef.current);
    }, []);
    const { ref: observerRef } = useMutationObserver(onMutation);
    const ref = (0, _compose_refs.composeRefs)(rootRef, observerRef);
    const onFocusCapture = _react.useCallback((e)=>{
        if (!(e.target instanceof HTMLElement)) return;
        focusedRef.current = e.target;
    }, []);
    const onBlur = _react.useCallback((e)=>{
        if (!(e.relatedTarget instanceof HTMLElement)) return;
        if (rootRef.current?.contains(e.relatedTarget)) return;
        focusedRef.current = null;
    }, []);
    const focusManager = _react.useMemo(()=>({
            get focused () {
                return focusedRef.current;
            },
            get focusables () {
                return focusablesRef.current.slice();
            },
            focusAt (index, options) {
                const { wrap = true } = options ?? {};
                const focusables = focusablesRef.current;
                const focusable = focusables[wrap ? index % focusables.length : index];
                if (!focusable) return;
                focusable.focus();
                return focusable;
            },
            focusNext (options) {
                const { wrap = true } = options ?? {};
                const focusables = focusablesRef.current;
                const currentIndex = focusables.findIndex((el)=>el === focusedRef.current);
                if (currentIndex === -1) return;
                const nextIndex = wrap ? (currentIndex + 1) % focusables.length : Math.min(currentIndex + 1, focusables.length - 1);
                const focusable = focusables[nextIndex];
                if (!focusable) return;
                focusable.focus();
                return focusable;
            },
            focusPrevious (options) {
                const { wrap = true } = options ?? {};
                const focusables = focusablesRef.current;
                const currentIndex = focusables.findIndex((el)=>el === focusedRef.current);
                if (currentIndex === -1) return;
                const prevIndex = wrap ? currentIndex > 0 ? currentIndex - 1 : focusables.length - 1 : Math.max(0, currentIndex - 1);
                const focusable = focusables[prevIndex];
                if (!focusable) return;
                focusable.focus();
                return focusable;
            },
            focusFirst () {
                const focusables = focusablesRef.current;
                const focusable = focusables[0];
                if (!focusable) return;
                focusable.focus();
                return focusable;
            },
            focusLast () {
                const focusables = focusablesRef.current;
                const focusable = focusables[focusables.length - 1];
                if (!focusable) return;
                focusable.focus();
                return focusable;
            }
        }), []);
    return {
        ref,
        onFocusCapture,
        onBlur,
        focusManager
    };
}
const FocusManagerContext = _react.createContext(null);
const FocusManagerProvider = FocusManagerContext.Provider;
function WithFocusManager(props) {
    const { ref, onFocusCapture, onBlur, focusManager } = useCreateFocusManager();
    return (0, _jsxruntime.jsx)(FocusManagerProvider, {
        value: focusManager,
        children: (0, _jsxruntime.jsx)("div", {
            ref: ref,
            onFocusCapture: onFocusCapture,
            onBlur: onBlur,
            className: _focuscss.default.nolayout,
            children: props.children
        })
    });
}
function useFocusManager() {
    return _react.useContext(FocusManagerContext);
}
