// @canva/app-ui-kit/dist/cjs/ui/base/alert/private/alert.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Alert () {
        return Alert;
    },
    get AlertIcon () {
        return AlertIcon;
    },
    get alertTones () {
        return alertTones;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _compose_refs = require('../../../../base/react/compose_refs');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _announcer = require('../../a11y/announcer/announcer');
const _box = require('../../box/box');
const _button = require('../../button/button');
const _focusable = require('../../focusable/focusable');
const _icon = require('../../icons/alert_triangle/icon');
const _icon1 = require('../../icons/check/icon');
const _icon2 = require('../../icons/info/icon');
const _icon3 = require('../../icons/x/icon');
const _layout = require('../../layout/layout');
const _linkcss = require('../../link/internal/link.css');
const _color = require('../../tokens/color');
const _typography = require('../../typography/typography');
const _alertcss = _interop_require_default(require("./alert.css"));
const _alertmessages = require("./alert.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const alertTones = [
    'positive',
    'warn',
    'critical',
    'info',
    'neutral'
];
const toneColors = {
    ['positive']: {
        fg: _color.colorFeedbackPositiveSubtleFg,
        bg: 'feedbackPositiveSubtle'
    },
    ['warn']: {
        fg: _color.colorFeedbackWarnSubtleFg,
        bg: 'feedbackWarnSubtle'
    },
    ['critical']: {
        fg: _color.colorFeedbackCriticalSubtleFg,
        bg: 'feedbackCriticalSubtle'
    },
    ['info']: {
        fg: _color.colorFeedbackInfoSubtleFg,
        bg: 'feedbackInfoSubtle'
    },
    ['neutral']: {
        fg: _color.colorActionSecondaryFg,
        bg: 'neutralSubtle'
    }
};
const AlertIcon = ({ tone, size = 'medium' })=>{
    switch(tone){
        case 'critical':
        case 'warn':
            return (0, _jsxruntime.jsx)(_icon.AlertTriangleIcon, {
                size: size
            });
        case 'positive':
            return (0, _jsxruntime.jsx)(_icon1.CheckIcon, {
                size: size
            });
        case 'info':
            return (0, _jsxruntime.jsx)(_icon2.InfoIcon, {
                size: size
            });
        case 'neutral':
            return null;
        default:
            throw new _preconditions.UnreachableError(tone);
    }
};
const Alert = ({ children, className, title, tone, dismissible, onDismiss, size = 'medium', rounded = true, showIcon = size !== 'small', Icon, focusRef, testId })=>{
    let iconElement;
    if (showIcon) {
        if (Icon) iconElement = (0, _jsxruntime.jsx)(Icon, {
            size: size
        });
        else iconElement = (0, _jsxruntime.jsx)(AlertIcon, {
            tone: tone,
            size: size
        });
    }
    const containerRef = (0, _focusable.useFocusableRef)(focusRef);
    const [textContent, setTextContent] = _react.useState('');
    const announcerRef = _react.useCallback((node)=>{
        if (node) setTextContent(node.textContent ?? '');
    }, []);
    (0, _announcer.useAnnounceWhen)(textContent, tone === 'critical' && !!textContent, {
        priority: 'high'
    });
    return (0, _jsxruntime.jsx)(_box.Box, {
        background: toneColors[tone].bg,
        borderRadius: rounded ? 'element' : 'none',
        paddingX: size === 'medium' ? '2u' : '1.5u',
        paddingY: size === 'medium' ? '2u' : '1u',
        width: "full",
        className: (0, _classnames.default)(className, _alertcss.default.alert, {
            [_alertcss.default.small]: size === 'small'
        }),
        testId: testId,
        customProperties: {
            [_linkcss.customProperties.linkColor]: toneColors[tone].fg
        },
        role: "note",
        ref: (0, _compose_refs.composeRefs)(containerRef, announcerRef),
        tabIndex: focusRef != null ? -1 : undefined,
        children: (0, _jsxruntime.jsx)(_typography.Text, {
            tagName: "div",
            size: size,
            tone: _typography.InheritColor,
            children: (0, _jsxruntime.jsxs)(_layout.Columns, {
                spacing: "1u",
                alignY: "baseline",
                children: [
                    iconElement && (0, _jsxruntime.jsx)(_layout.Column, {
                        width: "content",
                        children: iconElement
                    }),
                    (0, _jsxruntime.jsxs)(_layout.Column, {
                        children: [
                            title && (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
                                children: [
                                    (0, _jsxruntime.jsx)("strong", {
                                        children: title
                                    }),
                                    ' '
                                ]
                            }),
                            children
                        ]
                    }),
                    dismissible && (0, _jsxruntime.jsx)(_layout.Column, {
                        width: "content",
                        children: (0, _jsxruntime.jsx)(_button.Button, {
                            variant: "tertiary",
                            className: _alertcss.default.close,
                            onClick: onDismiss,
                            icon: _icon3.XIcon,
                            ariaLabel: _alertmessages.AlertMessages.dismiss(),
                            size: "small",
                            iconSize: size
                        })
                    })
                ]
            })
        })
    });
};
