// @canva/app-ui-kit/dist/esm/ui/base/color_picker/sat_val_picker/private/sat_val_picker.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { useIsMouseAccessMode } from '../../../a11y/access_mode_state/access_mode_state';
import { useIsTouchInput } from '../../../a11y/pointer_input_state/pointer_input_state';
import { HsvColor } from '../../hsv_color/hsv_color';
import { useColorPickerColor, useColorPickerControls } from '../../internal/color_picker_provider';
import { Handle } from '../../internal/handle';
import { DraggableView } from '../../../draggable/draggable_view';
import { useIsomorphicLayoutEffect } from '../../../hydration/use_isomorphic_layout_effect';
import styles, { customProperties } from './sat_val_picker.css';
export function SatValPicker(
    { color: colorProp, Draggable = DraggableView, onChange: onChangeProp, onChangeStart: onChangeStartProp, onChangeComplete: onChangeCompleteProp, showHandlePreview = 'touch-only' }
) {
    const colorPicker = useColorPickerControls();
    const setColor = colorPicker?.setColor;
    const onChangeStart = onChangeStartProp ?? colorPicker?.onChangeStart;
    const onChangeComplete = onChangeCompleteProp ?? colorPicker?.onChangeComplete;
    const colorPickerColor = useColorPickerColor();
    const color = colorProp ?? colorPickerColor ?? HsvColor.fromHexString('#ffffff');
    const pickerRef = React.useRef(null);
    const setPickerRef = React.useCallback((el)=>{
        pickerRef.current = el;
        el?.addEventListener('touchmove', (event)=>event.stopPropagation());
    }, []);
    const [isFocused, setIsFocused] = React.useState(false);
    const isMouseAccessMode = useIsMouseAccessMode();
    const [showPreview, setShowPreview] = React.useState(false);
    const isTouchInput = useIsTouchInput();
    const [browserWindow, setBrowserWindow] = React.useState();
    useIsomorphicLayoutEffect(()=>{
        const targetWindow = pickerRef.current?.ownerDocument.defaultView ?? window;
        setBrowserWindow({
            requestAnimationFrame: targetWindow.requestAnimationFrame,
            cancelAnimationFrame: targetWindow.cancelAnimationFrame,
            document: targetWindow.document
        });
    }, []);
    const [changing, setChanging] = React.useState(false);
    const onChange = ({ s, v })=>{
        if (!changing) {
            setChanging(true);
            onChangeStart && onChangeStart();
        }
        if (s !== color.s || v !== color.v) {
            setColor?.({
                s,
                v
            }, 'satVal');
            onChangeProp?.(color.with({
                s,
                v
            }));
        }
    };
    const moveHandle = (handle)=>{
        onChange({
            s: handle.leftPercent,
            v: 1 - handle.topPercent
        });
    };
    const stopChanging = React.useCallback(()=>{
        setChanging(false);
        setShowPreview(false);
        if (onChangeComplete) onChangeComplete();
    }, [
        onChangeComplete
    ]);
    const onKeyDown = (event)=>{
        const INCREMENT = 0.01;
        let { s, v } = color;
        switch(event.key){
            case 'ArrowUp':
                v += INCREMENT;
                break;
            case 'ArrowDown':
                v -= INCREMENT;
                break;
            case 'ArrowLeft':
                s -= INCREMENT;
                break;
            case 'ArrowRight':
                s += INCREMENT;
                break;
            default:
                return;
        }
        event.preventDefault();
        const clamp = (n)=>Math.min(Math.max(0, n), 1);
        onChange({
            s: clamp(s),
            v: clamp(v)
        });
    };
    const onKeyUp = React.useCallback(()=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const onMouseUp = React.useCallback(()=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const onMouseDown = (event)=>{
        if (event.button !== 0)
            return;
        pickerRef.current?.focus();
        moveHandle(mouseToHandleCoords(pickerRef.current.getBoundingClientRect(), event));
    };
    const onDragMove = (event)=>{
        moveHandle(mouseToHandleCoords(pickerRef.current.getBoundingClientRect(), event));
        if (showHandlePreview === 'always' || showHandlePreview === 'touch-only' && isTouchInput) setShowPreview(true);
    };
    const onDragEnd = React.useCallback((event)=>{
        stopChanging();
    }, [
        stopChanging
    ]);
    const handlePosition = {
        left: `${color.s * 100}%`,
        top: `${100 - color.v * 100}%`
    };
    const rootClassName = classNames(styles.root, {
        [styles.hideOutline]: isMouseAccessMode
    });
    const swatchColor = color.toRgb().toHexString();
    return _jsx(Draggable, {
        onMouseDown: onMouseDown,
        onDragMove: onDragMove,
        onDragEnd: onDragEnd,
        browserWindow: browserWindow,
        raf: browserWindow?.requestAnimationFrame,
        caf: browserWindow?.cancelAnimationFrame,
        children: ({
            onMouseDown,
            onTouchStart
        })=>
            _jsx("div", {
                className: rootClassName,
                onMouseDown: onMouseDown,
                onTouchStart: onTouchStart,
                onMouseUp: onMouseUp,
                onKeyDown: onKeyDown,
                onKeyUp: onKeyUp,
                onFocus: ()=>setIsFocused(true),
                onBlur: ()=>setIsFocused(false),
                ref: setPickerRef,
                tabIndex: 0,
                style: {
                    [customProperties.currentHue]: color.h * 360
                },
                children: _jsx(Handle, {
                    color: swatchColor,
                    active: changing,
                    focused: isFocused && !isMouseAccessMode,
                    style: handlePosition,
                    className: styles.handle,
                    preview: showPreview
                })
            })
    });
}
export function mouseToHandleCoords(pickerRect, { clientX, clientY }) {
    const clamp = (val)=>Math.min(1, Math.max(0, val));
    return {
        leftPercent: clamp((clientX - pickerRect.left) / pickerRect.width),
        topPercent: clamp((clientY - pickerRect.top) / pickerRect.height)
    };
}
