// @canva/app-ui-kit/dist/esm/ui/base/layer/private/layer.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { useInteractionLayer } from '../../a11y/interaction_manager/interaction_manager';
import { useIsomorphicLayoutEffect } from '../../hydration/use_isomorphic_layout_effect';
import { WithThemeData } from '../../theme/theme';
import { insertLayerIntoParent, LAYER_MARKER_CLASS_NAME } from './internal';
import styles from './layer.css';
export const LayerLevel = {
  CONTENT: 0,
  PINS: 1,
  LIGHTBOXES: 2,
  MODALS: 3,
  COMPLIANCE_NOTICES: 4,
  TOASTS: 5,
  TOOLTIPS: 6
};
export class LayerError extends Error {}
const LayerContext = React.createContext({});
export function WithLayerParent(
  {
    parentLayer,
    children
  }
) {
  return _jsx(LayerContext.Provider, {
    value: {
      parentLayer
    },
    children: children
  });
}
export const Layer = React.memo(function Layer({
  ...props
}) {
  const ownWindow = typeof window !== 'undefined' ? window : undefined;
  return ownWindow ? _jsx(LayerImpl, {
    ...props,
    ownWindow: ownWindow
  }) : null;
});
const LayerImpl = React.memo(function LayerImpl({
  ownWindow,
  open,
  level = 1,
  onOutsideLayerPointerDown,
  parentLayer: propsParentLayer,
  markOutsideInert = false,
  children
}) {
  const interactionLayerRef = useInteractionLayer({
    level,
    markOutsideInert
  });
  const context = React.useContext(LayerContext);
  const parentLayer = propsParentLayer || context.parentLayer || ownWindow.document.body;
  const parentDocument = parentLayer?.ownerDocument || ownWindow.document;
  const {
    layerRoot,
    layerContent
  } = React.useMemo(() => {
    if (!open) return {
      layerRoot: undefined,
      layerContent: undefined
    };
    const layerContent = parentDocument.createElement('div');
    layerContent.classList.add(styles.layerContent);
    const layerRoot = parentDocument.createElement('div');
    layerRoot.classList.add(LAYER_MARKER_CLASS_NAME);
    layerRoot.classList.add(styles.layerRoot);
    layerRoot.appendChild(layerContent);
    return {
      layerRoot,
      layerContent
    };
  }, [open, parentDocument]);
  React.useEffect(() => {
    if (layerRoot && parentLayer === parentDocument.body) {
      const noop = () => 0;
      layerRoot.addEventListener('click', noop);
      return () => {
        layerRoot?.removeEventListener('click', noop);
      };
    }
  }, [layerRoot, parentDocument, parentLayer]);
  const [mounted, setMounted] = React.useState(false);
  useIsomorphicLayoutEffect(() => {
    if (layerRoot && layerContent) {
      insertLayerIntoParent(layerRoot, parentLayer);
      interactionLayerRef(layerContent);
      setMounted(true);
      return () => {
        const parentGlobal = parentDocument?.defaultView || ownWindow;
        interactionLayerRef(null);
        parentLayer.removeChild(layerRoot);
        parentDocument?.dispatchEvent(new parentGlobal.CustomEvent('layerclose'));
        setMounted(false);
      };
    }
  }, [interactionLayerRef, layerContent, layerRoot, parentDocument, parentLayer, ownWindow]);
  useIsomorphicLayoutEffect(() => {
    if (layerRoot) layerRoot.style.zIndex = String(level);
  }, [layerRoot, level]);
  React.useEffect(() => {
    if (onOutsideLayerPointerDown == null || parentDocument == null) return;
    const handleDocumentMouseDown = event => {
      if (layerRoot == null)
        return;
      const parentGlobal = parentDocument?.defaultView || ownWindow;
      if (!(event.target instanceof parentGlobal.Node))
        return;
      if (!layerRoot.contains(event.target)) onOutsideLayerPointerDown(event);
    };
    parentDocument.addEventListener('mousedown', handleDocumentMouseDown);
    parentDocument.addEventListener('touchstart', handleDocumentMouseDown);
    return () => {
      parentDocument.removeEventListener('mousedown', handleDocumentMouseDown);
      parentDocument.removeEventListener('touchstart', handleDocumentMouseDown);
    };
  }, [layerRoot, onOutsideLayerPointerDown, parentDocument, ownWindow]);
  if (layerContent && mounted) return ReactDOM.createPortal(_jsx(LayerContext.Provider, {
    value: {
      parentLayer: layerRoot
    },
    children: _jsx(WithThemeData, {
      children: data => _jsx("div", {
        className: data.className,
        children: children
      })
    })
  }), layerContent);
  return null;
});
