// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/internal/color_picker_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get ColorPickerAlphaProvider () {
        return ColorPickerAlphaProvider;
    },
    get ColorPickerControlsProvider () {
        return ColorPickerControlsProvider;
    },
    get ColorPickerHueProvider () {
        return ColorPickerHueProvider;
    },
    get ColorPickerSaturationProvider () {
        return ColorPickerSaturationProvider;
    },
    get ColorPickerValueProvider () {
        return ColorPickerValueProvider;
    },
    get useColorPickerAlpha () {
        return useColorPickerAlpha;
    },
    get useColorPickerColor () {
        return useColorPickerColor;
    },
    get useColorPickerControls () {
        return useColorPickerControls;
    },
    get useColorPickerHex () {
        return useColorPickerHex;
    },
    get useColorPickerHue () {
        return useColorPickerHue;
    },
    get useColorPickerSaturation () {
        return useColorPickerSaturation;
    },
    get useColorPickerValue () {
        return useColorPickerValue;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _hsv_color = require('../hsv_color/hsv_color');
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ColorPickerHueContext = _react.createContext(undefined);
const ColorPickerSaturationContext = _react.createContext(undefined);
const ColorPickerValueContext = _react.createContext(undefined);
const ColorPickerAlphaContext = _react.createContext(undefined);
const ColorPickerControlsContext = _react.createContext(undefined);
function ColorPickerHueProvider({ hue, children }) {
    return (0, _jsxruntime.jsx)(ColorPickerHueContext.Provider, {
        value: hue,
        children: children
    });
}
function ColorPickerSaturationProvider({ saturation, children }) {
    return (0, _jsxruntime.jsx)(ColorPickerSaturationContext.Provider, {
        value: saturation,
        children: children
    });
}
function ColorPickerValueProvider({ value, children }) {
    return (0, _jsxruntime.jsx)(ColorPickerValueContext.Provider, {
        value: value,
        children: children
    });
}
function ColorPickerAlphaProvider({ alpha, children }) {
    return (0, _jsxruntime.jsx)(ColorPickerAlphaContext.Provider, {
        value: alpha ?? 1,
        children: children
    });
}
function ColorPickerControlsProvider({ setColor, onChangeStart, onChangeComplete, children }) {
    const value = _react.useMemo(()=>({
            setColor,
            onChangeStart,
            onChangeComplete
        }), [
        setColor,
        onChangeStart,
        onChangeComplete
    ]);
    return (0, _jsxruntime.jsx)(ColorPickerControlsContext.Provider, {
        value: value,
        children: children
    });
}
function useColorPickerHue() {
    return _react.useContext(ColorPickerHueContext);
}
function useColorPickerSaturation() {
    return _react.useContext(ColorPickerSaturationContext);
}
function useColorPickerValue() {
    return _react.useContext(ColorPickerValueContext);
}
function useColorPickerAlpha() {
    return _react.useContext(ColorPickerAlphaContext);
}
function useColorPickerColor() {
    const h = useColorPickerHue();
    const s = useColorPickerSaturation();
    const v = useColorPickerValue();
    const a = useColorPickerAlpha();
    if (h == null || s == null || v == null) return undefined;
    return new _hsv_color.HsvColor(h, s, v, a);
}
function useColorPickerHex() {
    const color = useColorPickerColor();
    return color?.toRgb().toHexString();
}
function useColorPickerControls() {
    return _react.useContext(ColorPickerControlsContext);
}
