// @canva/app-ui-kit/dist/esm/ui/base/surface/sheet/private/sheet_wrapper.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { CSSTransition } from 'react-transition-group';
import { Layer } from '../../../layer/layer';
import { useEnableAnimations } from '../../../provider/provider';
import { useStableFunction } from '../../../stable_function/stable_function';
import styles from './sheet.css';
export function SheetWrapper(props) {
    const Wrapper = useEnableAnimations() ? TransitionWrapper : StaticWrapper;
    return _jsx(Wrapper, {
        ...props
    });
}
const TRANSITION_DURATION = parseInt(styles.transitionTime, 10);
const TRANSITION_CLASS_NAMES = {
    appear: styles.transitioning,
    enter: styles.transitioning,
    exit: styles.transitioning,
    appearActive: styles.visible,
    appearDone: styles.visible,
    enterActive: styles.visible,
    enterDone: styles.visible
};
function TransitionWrapper(
    { open: propOpen, onOpenComplete, onCloseComplete, markOutsideInert, children }
) {
    const [transitionOpen, setTransitionOpen] = React.useState(propOpen);
    const nodeRef = React.useRef(null);
    return _jsx(Layer, {
        open: propOpen || transitionOpen,
        markOutsideInert: markOutsideInert,
        children: _jsx(CSSTransition, {
            in: propOpen,
            appear: true,
            timeout: TRANSITION_DURATION,
            classNames: TRANSITION_CLASS_NAMES,
            onEnter: ()=>setTransitionOpen(true),
            onEntered: ()=>onOpenComplete?.(),
            onExited: ()=>{
                setTransitionOpen(false);
                onCloseComplete?.();
            },
            nodeRef: nodeRef,
            children: _jsx("div", {
                ref: nodeRef,
                children: children
            })
        })
    });
}
function StaticWrapper({ open, onCloseComplete, onOpenComplete, markOutsideInert, children }) {
    const prevOpen = React.useRef(false);
    const stableOnCloseComplete = useStableFunction(()=>onCloseComplete?.());
    const stableOnOpenComplete = useStableFunction(()=>onOpenComplete?.());
    React.useEffect(()=>{
        if (prevOpen.current === open)
        return;
        prevOpen.current = open;
        if (open) stableOnOpenComplete();
        else stableOnCloseComplete();
        prevOpen.current = open;
    }, [
        open,
        stableOnOpenComplete,
        stableOnCloseComplete
    ]);
    return _jsx(Layer, {
        open: open,
        markOutsideInert: markOutsideInert,
        children: _jsx("div", {
            className: styles.visible,
            children: children
        })
    });
}
