// @canva/app-ui-kit/dist/esm/ui/base/responsive/private/breakpoint_store.js
import { debounce } from '../../../../base/debounce';
import { Preconditions } from '../../../../base/preconditions';
import { Breakpoint } from '../../metrics/metrics';
import metricsStyles from '../../metrics/metrics.css';
const MEDIA_QUERIES = [
    metricsStyles.smallUp,
    metricsStyles.mediumUp,
    metricsStyles.largeUp,
    metricsStyles.xLargeUp
];
const BREAKPOINT_MAP = new Map([
    [
        0,
        Breakpoint.DEFAULT
    ],
    [
        1,
        Breakpoint.SMALL
    ],
    [
        2,
        Breakpoint.MEDIUM
    ],
    [
        3,
        Breakpoint.LARGE
    ],
    [
        4,
        Breakpoint.XLARGE
    ]
]);
const DEFAULT_BREAKPOINT_FOR_TESTS = Breakpoint.DEFAULT;
class BreakpointStore {
    setSsrBreakpoint(breakpoint) {
        this.serverBreakpoint = breakpoint;
    }
    setTestBreakpoint(breakpoint) {
        this.testBreakpoint = breakpoint;
        this.notifySubscribers();
    }
    resetConfiguration() {
        this.serverBreakpoint = undefined;
        this.testBreakpoint = undefined;
    }
    notifySubscribers() {
        this.subscribers.forEach((callback)=>callback());
    }
    constructor(){
        this.subscribers = new Set();
        this.subscribe = (callback)=>{
            this.subscribers.add(callback);
            return ()=>this.subscribers.delete(callback);
        };
        this.getSnapshot = ()=>this.clientBreakpoint ?? this.testBreakpoint ?? DEFAULT_BREAKPOINT_FOR_TESTS;
        this.getServerSnapshot = ()=>Preconditions.checkExists(this.serverBreakpoint, 'SSR breakpoint is not defined. Call setSsrBreakpoint() to configure a breakpoint on SSR pages.');
        this.update = ()=>{
            if (this.mqls) {
                const matchCount = this.mqls.filter((mql)=>mql.matches).length;
                this.clientBreakpoint = BREAKPOINT_MAP.get(matchCount);
                this.notifySubscribers();
            }
        };
        this.listener = debounce(this.update, 16);
        const matchMedia = typeof window !== 'undefined' && window.matchMedia;
        if (matchMedia) {
            this.mqls = MEDIA_QUERIES.map((query)=>matchMedia(query));
            this.mqls.forEach((mql)=>safeAddListener(mql, this.listener));
            this.update();
        }
    }
}
export const breakpointStore = new BreakpointStore();
function safeAddListener(mql, listener) {
    if (typeof mql.addEventListener === 'function') mql.addEventListener('change', listener);
    else mql.addListener(listener);
}
