// @canva/app-ui-kit/dist/cjs/ui/base/device_capabilities/keyboard.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Internal () {
        return Internal;
    },
    get applyKeyboardViewportOffsetCompensation () {
        return applyKeyboardViewportOffsetCompensation;
    },
    get getNonInteractiveHeight () {
        return getNonInteractiveHeight;
    },
    get getVirtualKeyboardOverlayHeight () {
        return getVirtualKeyboardOverlayHeight;
    },
    get guessKeyboardOverlayHeight () {
        return guessKeyboardOverlayHeight;
    },
    get isVirtualKeyboard () {
        return isVirtualKeyboard;
    },
    get maybeWaitForVirtualKeyboard () {
        return maybeWaitForVirtualKeyboard;
    },
    get needsIosSafari26KeyboardAdditionalHeight () {
        return needsIosSafari26KeyboardAdditionalHeight;
    },
    get useIsVirtualKeyboard () {
        return useIsVirtualKeyboard;
    }
});
const _delay = require('../../../base/delay');
const _react = require("react");
const _visual_viewport = require("./visual_viewport");
const globalNavigator = typeof navigator !== 'undefined' ? navigator : undefined;
function useIsVirtualKeyboard() {
    return (0, _react.useSyncExternalStore)(subscribeIsVirtualKeyboard, getClientIsVirtualKeyboard, getServerIsVirtualKeyboard);
}
function isVirtualKeyboard(navigator1 = globalNavigator) {
    return navigator1 != null && !!navigator1.userAgent.match(/iphone|ipad|android/gi);
}
function getVirtualKeyboardOverlayHeight() {
    const visualViewport = (0, _visual_viewport.getVisualViewport)();
    const visualViewportHeight = visualViewport?.height ?? 0;
    const viewportHeight = Math.max(window.document.documentElement.clientHeight, window.innerHeight);
    const baseVirtualKeyboardHeight = Math.max(viewportHeight - visualViewportHeight, 0);
    return baseVirtualKeyboardHeight === 0 ? baseVirtualKeyboardHeight : baseVirtualKeyboardHeight + getNonInteractiveHeight();
}
function needsIosSafari26KeyboardAdditionalHeight() {
    return Internal.isIosSafari26OrAbove() && !Internal.isStandalone();
}
function guessKeyboardOverlayHeight(navigator1 = globalNavigator) {
    if (!isVirtualKeyboard(navigator1)) return 0;
    if (window.innerHeight < window.screen.height - 240)
        return 0;
    return Math.max(270, Math.min(340, window.screen.height * 0.45))
     + (Internal.isCanvaEditor(navigator1) ? 80 : 0);
}
const Internal = {
    isCanvaEditor(navigator1 = globalNavigator) {
        return navigator1 != null && navigator1.userAgent.indexOf('canvaeditor') >= 0;
    },
    isIos26OrAbove(_navigator = globalNavigator) {
        return _navigator != null && !!_navigator.userAgent.match(/iPhone OS 18_/gi) && !_navigator.userAgent.match(/Version\/18/gi);
    },
    isIos26_0(_navigator = globalNavigator) {
        return _navigator != null && Internal.isIos26OrAbove(_navigator) && !!_navigator.userAgent.match(/Version\/26\.0/gi);
    },
    isIosSafari26OrAbove(navigator1 = globalNavigator) {
        return !Internal.isCanvaEditor(navigator1) && Internal.isIos26OrAbove(navigator1);
    },
    isStandalone(navigator1 = globalNavigator) {
        return navigator1 != null && !!navigator1.standalone;
    }
};
const getNonInteractiveHeight = ()=>{
    if (!Internal.isIosSafari26OrAbove() || Internal.isStandalone())
        return 0;
    return Internal.isIos26_0() ? 36 : 20;
};
function applyKeyboardViewportOffsetCompensation(height) {
    const visualViewport = (0, _visual_viewport.getVisualViewport)();
    const visualViewportOffsetTop = visualViewport?.offsetTop ?? 0;
    return Math.max(0, height - visualViewportOffsetTop);
}
function maybeWaitForVirtualKeyboard(maxWaitTimeMs = 300) {
    const visualViewport = (0, _visual_viewport.getVisualViewport)();
    if (!isVirtualKeyboard() || !visualViewport || getVirtualKeyboardOverlayHeight() > 0) return (0, _delay.delay)(0);
    return new Promise((resolve)=>{
        const removeEventListenerAndResolve = ()=>{
            visualViewport.removeEventListener('resize', removeEventListenerAndResolve);
            resolve();
        };
        visualViewport.addEventListener('resize', removeEventListenerAndResolve);
        (0, _delay.delay)(maxWaitTimeMs).then(removeEventListenerAndResolve);
    });
}
const subscribeIsVirtualKeyboard = (_input)=>()=>{};
const isVirtualKeyboardValue = isVirtualKeyboard();
const getClientIsVirtualKeyboard = ()=>isVirtualKeyboardValue;
const getServerIsVirtualKeyboard = ()=>false;
