// @canva/app-ui-kit/dist/cjs/ui/base/tokens/primitive/color.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get colorBlackA00 () {
        return colorBlackA00;
    },
    get colorBlackA01 () {
        return colorBlackA01;
    },
    get colorBlackA02 () {
        return colorBlackA02;
    },
    get colorBlackA03 () {
        return colorBlackA03;
    },
    get colorBlackA04 () {
        return colorBlackA04;
    },
    get colorBlackA05 () {
        return colorBlackA05;
    },
    get colorBlackA06 () {
        return colorBlackA06;
    },
    get colorBlackA07 () {
        return colorBlackA07;
    },
    get colorBlackA08 () {
        return colorBlackA08;
    },
    get colorBlackA09 () {
        return colorBlackA09;
    },
    get colorBlackA10 () {
        return colorBlackA10;
    },
    get colorBlue01 () {
        return colorBlue01;
    },
    get colorBlue02 () {
        return colorBlue02;
    },
    get colorBlue03 () {
        return colorBlue03;
    },
    get colorBlue04 () {
        return colorBlue04;
    },
    get colorBlue05 () {
        return colorBlue05;
    },
    get colorBlue06 () {
        return colorBlue06;
    },
    get colorBlue07 () {
        return colorBlue07;
    },
    get colorBlue08 () {
        return colorBlue08;
    },
    get colorBlue09 () {
        return colorBlue09;
    },
    get colorBlue10 () {
        return colorBlue10;
    },
    get colorGreen01 () {
        return colorGreen01;
    },
    get colorGreen02 () {
        return colorGreen02;
    },
    get colorGreen03 () {
        return colorGreen03;
    },
    get colorGreen04 () {
        return colorGreen04;
    },
    get colorGreen05 () {
        return colorGreen05;
    },
    get colorGreen06 () {
        return colorGreen06;
    },
    get colorGreen07 () {
        return colorGreen07;
    },
    get colorGreen08 () {
        return colorGreen08;
    },
    get colorGreen09 () {
        return colorGreen09;
    },
    get colorGreen10 () {
        return colorGreen10;
    },
    get colorGrey01 () {
        return colorGrey01;
    },
    get colorGrey02 () {
        return colorGrey02;
    },
    get colorGrey03 () {
        return colorGrey03;
    },
    get colorGrey04 () {
        return colorGrey04;
    },
    get colorGrey05 () {
        return colorGrey05;
    },
    get colorGrey06 () {
        return colorGrey06;
    },
    get colorGrey07 () {
        return colorGrey07;
    },
    get colorGrey08 () {
        return colorGrey08;
    },
    get colorGrey09 () {
        return colorGrey09;
    },
    get colorGrey10 () {
        return colorGrey10;
    },
    get colorPurple01 () {
        return colorPurple01;
    },
    get colorPurple02 () {
        return colorPurple02;
    },
    get colorPurple03 () {
        return colorPurple03;
    },
    get colorPurple04 () {
        return colorPurple04;
    },
    get colorPurple05 () {
        return colorPurple05;
    },
    get colorPurple06 () {
        return colorPurple06;
    },
    get colorPurple07 () {
        return colorPurple07;
    },
    get colorPurple08 () {
        return colorPurple08;
    },
    get colorPurple09 () {
        return colorPurple09;
    },
    get colorPurple10 () {
        return colorPurple10;
    },
    get colorPurpleA00 () {
        return colorPurpleA00;
    },
    get colorPurpleA01 () {
        return colorPurpleA01;
    },
    get colorPurpleA02 () {
        return colorPurpleA02;
    },
    get colorPurpleA03 () {
        return colorPurpleA03;
    },
    get colorPurpleA04 () {
        return colorPurpleA04;
    },
    get colorPurpleA05 () {
        return colorPurpleA05;
    },
    get colorPurpleA06 () {
        return colorPurpleA06;
    },
    get colorPurpleA07 () {
        return colorPurpleA07;
    },
    get colorPurpleA08 () {
        return colorPurpleA08;
    },
    get colorPurpleA09 () {
        return colorPurpleA09;
    },
    get colorPurpleA10 () {
        return colorPurpleA10;
    },
    get colorRed01 () {
        return colorRed01;
    },
    get colorRed02 () {
        return colorRed02;
    },
    get colorRed03 () {
        return colorRed03;
    },
    get colorRed04 () {
        return colorRed04;
    },
    get colorRed05 () {
        return colorRed05;
    },
    get colorRed06 () {
        return colorRed06;
    },
    get colorRed07 () {
        return colorRed07;
    },
    get colorRed08 () {
        return colorRed08;
    },
    get colorRed09 () {
        return colorRed09;
    },
    get colorRed10 () {
        return colorRed10;
    },
    get colorTeal01 () {
        return colorTeal01;
    },
    get colorTeal02 () {
        return colorTeal02;
    },
    get colorTeal03 () {
        return colorTeal03;
    },
    get colorTeal04 () {
        return colorTeal04;
    },
    get colorWhiteA00 () {
        return colorWhiteA00;
    },
    get colorWhiteA01 () {
        return colorWhiteA01;
    },
    get colorWhiteA02 () {
        return colorWhiteA02;
    },
    get colorWhiteA03 () {
        return colorWhiteA03;
    },
    get colorWhiteA04 () {
        return colorWhiteA04;
    },
    get colorWhiteA05 () {
        return colorWhiteA05;
    },
    get colorWhiteA06 () {
        return colorWhiteA06;
    },
    get colorWhiteA07 () {
        return colorWhiteA07;
    },
    get colorWhiteA08 () {
        return colorWhiteA08;
    },
    get colorWhiteA09 () {
        return colorWhiteA09;
    },
    get colorWhiteA10 () {
        return colorWhiteA10;
    },
    get colorYellow01 () {
        return colorYellow01;
    },
    get colorYellow02 () {
        return colorYellow02;
    },
    get colorYellow03 () {
        return colorYellow03;
    },
    get colorYellow04 () {
        return colorYellow04;
    },
    get colorYellow05 () {
        return colorYellow05;
    },
    get colorYellow06 () {
        return colorYellow06;
    },
    get colorYellow07 () {
        return colorYellow07;
    },
    get colorYellow08 () {
        return colorYellow08;
    },
    get colorYellow09 () {
        return colorYellow09;
    },
    get colorYellow10 () {
        return colorYellow10;
    }
});
const _colorcss = _interop_require_default(require("./private/color.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const colorTeal01 = _colorcss.default.colorTeal01;
const colorTeal02 = _colorcss.default.colorTeal02;
const colorTeal03 = _colorcss.default.colorTeal03;
const colorTeal04 = _colorcss.default.colorTeal04;
const colorGreen01 = _colorcss.default.colorGreen01;
const colorGreen02 = _colorcss.default.colorGreen02;
const colorGreen03 = _colorcss.default.colorGreen03;
const colorGreen04 = _colorcss.default.colorGreen04;
const colorGreen05 = _colorcss.default.colorGreen05;
const colorGreen06 = _colorcss.default.colorGreen06;
const colorGreen07 = _colorcss.default.colorGreen07;
const colorGreen08 = _colorcss.default.colorGreen08;
const colorGreen09 = _colorcss.default.colorGreen09;
const colorGreen10 = _colorcss.default.colorGreen10;
const colorPurple01 = _colorcss.default.colorPurple01;
const colorPurple02 = _colorcss.default.colorPurple02;
const colorPurple03 = _colorcss.default.colorPurple03;
const colorPurple04 = _colorcss.default.colorPurple04;
const colorPurple05 = _colorcss.default.colorPurple05;
const colorPurple06 = _colorcss.default.colorPurple06;
const colorPurple07 = _colorcss.default.colorPurple07;
const colorPurple08 = _colorcss.default.colorPurple08;
const colorPurple09 = _colorcss.default.colorPurple09;
const colorPurple10 = _colorcss.default.colorPurple10;
const colorPurpleA00 = _colorcss.default.colorPurpleA00;
const colorPurpleA01 = _colorcss.default.colorPurpleA01;
const colorPurpleA02 = _colorcss.default.colorPurpleA02;
const colorPurpleA03 = _colorcss.default.colorPurpleA03;
const colorPurpleA04 = _colorcss.default.colorPurpleA04;
const colorPurpleA05 = _colorcss.default.colorPurpleA05;
const colorPurpleA06 = _colorcss.default.colorPurpleA06;
const colorPurpleA07 = _colorcss.default.colorPurpleA07;
const colorPurpleA08 = _colorcss.default.colorPurpleA08;
const colorPurpleA09 = _colorcss.default.colorPurpleA09;
const colorPurpleA10 = _colorcss.default.colorPurpleA10;
const colorRed01 = _colorcss.default.colorRed01;
const colorRed02 = _colorcss.default.colorRed02;
const colorRed03 = _colorcss.default.colorRed03;
const colorRed04 = _colorcss.default.colorRed04;
const colorRed05 = _colorcss.default.colorRed05;
const colorRed06 = _colorcss.default.colorRed06;
const colorRed07 = _colorcss.default.colorRed07;
const colorRed08 = _colorcss.default.colorRed08;
const colorRed09 = _colorcss.default.colorRed09;
const colorRed10 = _colorcss.default.colorRed10;
const colorBlue01 = _colorcss.default.colorBlue01;
const colorBlue02 = _colorcss.default.colorBlue02;
const colorBlue03 = _colorcss.default.colorBlue03;
const colorBlue04 = _colorcss.default.colorBlue04;
const colorBlue05 = _colorcss.default.colorBlue05;
const colorBlue06 = _colorcss.default.colorBlue06;
const colorBlue07 = _colorcss.default.colorBlue07;
const colorBlue08 = _colorcss.default.colorBlue08;
const colorBlue09 = _colorcss.default.colorBlue09;
const colorBlue10 = _colorcss.default.colorBlue10;
const colorYellow01 = _colorcss.default.colorYellow01;
const colorYellow02 = _colorcss.default.colorYellow02;
const colorYellow03 = _colorcss.default.colorYellow03;
const colorYellow04 = _colorcss.default.colorYellow04;
const colorYellow05 = _colorcss.default.colorYellow05;
const colorYellow06 = _colorcss.default.colorYellow06;
const colorYellow07 = _colorcss.default.colorYellow07;
const colorYellow08 = _colorcss.default.colorYellow08;
const colorYellow09 = _colorcss.default.colorYellow09;
const colorYellow10 = _colorcss.default.colorYellow10;
const colorWhiteA00 = _colorcss.default.colorWhiteA00;
const colorWhiteA01 = _colorcss.default.colorWhiteA01;
const colorWhiteA02 = _colorcss.default.colorWhiteA02;
const colorWhiteA03 = _colorcss.default.colorWhiteA03;
const colorWhiteA04 = _colorcss.default.colorWhiteA04;
const colorWhiteA05 = _colorcss.default.colorWhiteA05;
const colorWhiteA06 = _colorcss.default.colorWhiteA06;
const colorWhiteA07 = _colorcss.default.colorWhiteA07;
const colorWhiteA08 = _colorcss.default.colorWhiteA08;
const colorWhiteA09 = _colorcss.default.colorWhiteA09;
const colorWhiteA10 = _colorcss.default.colorWhiteA10;
const colorGrey01 = _colorcss.default.colorGrey01;
const colorGrey02 = _colorcss.default.colorGrey02;
const colorGrey03 = _colorcss.default.colorGrey03;
const colorGrey04 = _colorcss.default.colorGrey04;
const colorGrey05 = _colorcss.default.colorGrey05;
const colorGrey06 = _colorcss.default.colorGrey06;
const colorGrey07 = _colorcss.default.colorGrey07;
const colorGrey08 = _colorcss.default.colorGrey08;
const colorGrey09 = _colorcss.default.colorGrey09;
const colorGrey10 = _colorcss.default.colorGrey10;
const colorBlackA00 = _colorcss.default.colorBlackA00;
const colorBlackA01 = _colorcss.default.colorBlackA01;
const colorBlackA02 = _colorcss.default.colorBlackA02;
const colorBlackA03 = _colorcss.default.colorBlackA03;
const colorBlackA04 = _colorcss.default.colorBlackA04;
const colorBlackA05 = _colorcss.default.colorBlackA05;
const colorBlackA06 = _colorcss.default.colorBlackA06;
const colorBlackA07 = _colorcss.default.colorBlackA07;
const colorBlackA08 = _colorcss.default.colorBlackA08;
const colorBlackA09 = _colorcss.default.colorBlackA09;
const colorBlackA10 = _colorcss.default.colorBlackA10;
