// @canva/app-ui-kit/dist/esm/ui/base/scroll/scrollable/private/scrollable.js
import { jsx as _jsx } from "react/jsx-runtime";
import 'react';
import { BaseScrollable } from './base_scrollable';
import { ScrollableWithIndicator } from './scrollable_with_indicator';
export const Scrollable = (props)=>{
    const { children, direction = 'vertical', onScroll, onScrollStateChange, debounceMs, indicator, ariaLabel, role, ref } = props;
    if (indicator) return _jsx(ScrollableWithIndicator, {
        direction: direction,
        onScroll: onScroll,
        onScrollStateChange: onScrollStateChange,
        debounceMs: debounceMs,
        ariaLabel: ariaLabel,
        role: role,
        ref: ref,
        type: indicator.type,
        background: indicator.background ? indicator.background : 'surface',
        children: children
    });
    return _jsx(BaseScrollable, {
        direction: direction,
        onScroll: onScroll,
        onScrollStateChange: onScrollStateChange,
        debounceMs: debounceMs,
        ariaLabel: ariaLabel,
        role: role,
        ref: ref,
        children: children
    });
};
