// @canva/app-ui-kit/dist/cjs/ui/base/button/private/button.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BasicButton () {
        return BasicButton;
    },
    get BasicButtonLink () {
        return BasicButtonLink;
    },
    get Button () {
        return Button;
    },
    get ButtonLink () {
        return ButtonLink;
    },
    get CircleButton () {
        return CircleButton;
    },
    get LinkButton () {
        return LinkButton;
    },
    get PrimaryButton () {
        return PrimaryButton;
    },
    get SecondaryButton () {
        return SecondaryButton;
    },
    get TertiaryButton () {
        return TertiaryButton;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _internal_button = require("./internal_button");
function Button(props) {
    return (0, _jsxruntime.jsx)(_internal_button.InternalButton, {
        ...props
    });
}
function ButtonLink(props) {
    return (0, _jsxruntime.jsx)(_internal_button.InternalAnchor, {
        ...props
    });
}
function BasicButton({ ref, ...props }) {
    return (0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'basic'
    }, ref);
}
function BasicButtonLink(props) {
    return (0, _jsxruntime.jsx)(_internal_button.InternalAnchor, {
        ...props,
        variant: "basic"
    });
}
const PrimaryButton = (props)=>(0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'primary'
    });
const SecondaryButton = (props)=>(0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'secondary'
    });
const TertiaryButton = (props)=>(0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'tertiary'
    });
const LinkButton = (props)=>(0, _internal_button.renderButtonOrAnchor)({
        ...props,
        variant: 'linkButton'
    });
function isCircleButtonWithChildren(props) {
    return typeof props.children !== 'undefined';
}
function CircleButton(props) {
    if (isCircleButtonWithChildren(props)) return (0, _jsxruntime.jsx)(_internal_button.CircleButtonWithLabel, {
        ...props
    });
    return (0, _jsxruntime.jsx)(_internal_button.CircleButtonWithoutLabel, {
        ...props
    });
}
