// @canva/app-ui-kit/dist/cjs/ui/base/card/private/card.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Card () {
        return Card;
    },
    get SelectableWrapper () {
        return SelectableWrapper;
    },
    get WithCardState () {
        return WithCardState;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _classnames = _interop_require_default(require("classnames"));
const _mobxreactlite = require("mobx-react-lite");
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../../a11y/access_mode_state/access_mode_state');
const _badge = require('../../badge/badge');
const _temporary_exports = require('../../badge/temporary_exports');
const _layout = require('../../layout/layout');
const _provider = require('../../provider/provider');
const _thumbnail = require('../../thumbnail/thumbnail');
const _cardcss = _interop_require_default(require("./card.css"));
const _card_content = require("./card_content");
const _card_context = require("./card_context");
const _card_thumbnail = require("./card_thumbnail");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Card = (0, _card_context.withCardContextProvider)((0, _mobxreactlite.observer)(function Card({ title, description, preTitle, content, ariaControls, ariaHasPopup, ariaExpanded, ariaLabel, ariaLabelledBy, buttonAriaLabel, buttonAriaLabelledBy, href, target, draggable, buttonId, buttonRef, onClick, onDragStart, onDragEnd, onContextMenu, tabIndex, disabled = false, border = 'none', borderRadius = 'element', bleedX, padding = '0', background = 'none', layout = 'vertical', thumbnail: thumbnailProp, stretchThumbnail, decorators, decoratorsInset = '1u', decoratorsBoundary: decoratorsBoundaryProp, hoverEffect = 'none', hoverBoundary: hoverBoundaryProp, selectableMode = 'none', selected = false, selectableBoundary: selectableBoundaryProp, onPointerDown, onPointerUp, onPointerEnter, onPointerLeave, onPointerMove, onPointerCancel }) {
    const { decoratorsBoundary, hoverBoundary, selectableBoundary } = getBoundaries({
        decoratorsBoundary: decoratorsBoundaryProp,
        hoverBoundary: hoverBoundaryProp,
        selectableBoundary: selectableBoundaryProp
    });
    (0, _card_context.useCardConfigurationSetter)({
        disabled,
        layout,
        border,
        borderRadius,
        hoverEffect,
        hoverBoundary,
        selected,
        selectableMode,
        selectableBoundary
    });
    const enableAnimations = (0, _provider.useEnableAnimations)();
    const { onMouseEnter, onMouseLeave, onMouseUp, onMouseDown } = (0, _card_context.useCardEventHandlers)();
    const titleId = _react.useId();
    const descriptionId = _react.useId();
    const hasContent = !!(title || description || preTitle || content);
    const contentIds = hasContent ? {
        titleId,
        descriptionId
    } : undefined;
    const cardContainerIsGroup = decorators != null;
    const cardContainerAriaProps = getCardContainerAriaProps({
        ariaLabelledBy,
        ariaLabel,
        cardContainerIsGroup,
        contentIds
    });
    const buttonAriaProps = getButtonAriaProps({
        ariaLabelledBy,
        ariaLabel,
        buttonAriaLabelledBy,
        buttonAriaLabel,
        cardContainerIsGroup,
        contentIds
    });
    const gutterSize = {
        ['horizontal']: '1.5u',
        ['vertical']: '1u'
    };
    const isThumbnailOnlyCard = decoratorsBoundary === 'thumbnail' && !hasContent && padding === '0';
    const cardContainerClassName = (0, _classnames.default)(_cardcss.default.cardContainer, {
        [_cardcss.default.disabled]: disabled,
        [_cardcss.default.animated]: enableAnimations,
        [_cardcss.default.cardHoverTrigger]: selectableMode === 'primary',
        [_cardcss.default.scaleOnHover]: hoverEffect === 'scaleThumbnail',
        [_cardcss.default.bgTransitionOnHover]: hoverEffect === 'bgColorTransition' && hoverBoundary === 'card'
    });
    const cardContainerProps = {
        onMouseEnter,
        onMouseLeave,
        onMouseUp,
        onMouseDown,
        ...cardContainerAriaProps
    };
    const cardButton = (0, _jsxruntime.jsx)(InternalCardButton, {
        href: href,
        target: target,
        onClick: onClick,
        onPointerDown: onPointerDown,
        onPointerUp: onPointerUp,
        onPointerEnter: onPointerEnter,
        onPointerLeave: onPointerLeave,
        onPointerMove: onPointerMove,
        onPointerCancel: onPointerCancel,
        buttonAriaLabel: buttonAriaProps.buttonAriaLabel,
        buttonAriaLabelledBy: buttonAriaProps.buttonAriaLabelledBy,
        buttonAriaDescribedBy: buttonAriaProps.buttonAriaDescribedBy,
        ariaPressed: selectableMode === 'primary' ? selected || false : undefined,
        ariaControls: ariaControls,
        ariaHasPopup: ariaHasPopup,
        ariaExpanded: ariaExpanded,
        buttonId: buttonId,
        buttonRef: buttonRef,
        disabled: disabled,
        draggable: draggable,
        onDragStart: onDragStart,
        onDragEnd: onDragEnd,
        onContextMenu: onContextMenu,
        tabIndex: tabIndex
    });
    const badgeContainerProps = {
        paddingStyle: decoratorsInset === '1u' ? 'standard' : 'compact'
    };
    const thumbnail = typeof thumbnailProp === 'string' ? (0, _jsxruntime.jsx)(_card_thumbnail.CardImageThumbnail, {
        src: thumbnailProp
    }) : thumbnailProp;
    const decoratedThumbnail = decoratorsBoundary === 'thumbnail' ? (0, _jsxruntime.jsxs)(_badge.BadgeContainer, {
        ...(isThumbnailOnlyCard ? cardContainerProps : {}),
        ...badgeContainerProps,
        className: (0, _classnames.default)(_cardcss.default.thumbnailArea, isThumbnailOnlyCard && cardContainerClassName, {
            [_cardcss.default.stretchThumbnail]: stretchThumbnail
        }),
        children: [
            isThumbnailOnlyCard && cardButton,
            thumbnail,
            decorators
        ]
    }) : layout === 'horizontal' ? (0, _jsxruntime.jsx)("div", {
        className: _cardcss.default.thumbnailArea,
        children: thumbnail
    }) : thumbnail;
    if (isThumbnailOnlyCard) return decoratedThumbnail;
    const CardWithContentContainer = decoratorsBoundary === 'card' ? _badge.BadgeContainer : _temporary_exports.ExtendedBadgeVisibilityTrigger;
    return ((0, _jsxruntime.jsx)(SelectableWrapper, {
                boundary: "card",
                borderRadius: borderRadius,
                className: (0, _classnames.default)({
                    [_cardcss.default.horizontalWrapper]: layout === 'horizontal'
                }),
                children: (0, _jsxruntime.jsxs)(CardWithContentContainer, {
                    ...cardContainerProps,
                    ...(decoratorsBoundary === 'card' ? badgeContainerProps : {}),
                    className: (0, _classnames.default)(cardContainerClassName, {
                        [_cardcss.default.backgroundNeutralLow]: background === 'neutralLow',
                        [_cardcss.default.backgroundNone]: background === 'none',
                        [_cardcss.default.borderLow]: border === 'low',
                        [_cardcss.default.borderRadiusElement]: borderRadius === 'element',
                        [_cardcss.default.borderRadiusElementSmall]: borderRadius === 'elementSmall',
                        [_cardcss.default.borderRadiusElementRelaxed]: borderRadius === 'elementRelaxed',
                        [_cardcss.default.borderRadiusElementSoftest]: borderRadius === 'elementSoftest',
                        [_cardcss.default.paddingOneUnit]: padding === '1u',
                        [_cardcss.default.paddingTwoUnit]: padding === '2u',
                        [_cardcss.default.bleedXOneUnit]: bleedX && padding === '1u',
                        [_cardcss.default.bleedXTwoUnit]: bleedX && padding === '2u',
                        [_cardcss.default.vertical]: layout === 'vertical',
                        [_cardcss.default.horizontal]: layout === 'horizontal',
                        [_cardcss.default.selected]: selected
                    }),
                    children: [
                        cardButton,
                        thumbnail != null && (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
                            children: [
                                decoratedThumbnail,
                                (0, _jsxruntime.jsx)(_layout.Spacer, {
                                    size: gutterSize[layout],
                                    direction: layout
                                })
                            ]
                        }),
                        hasContent && (0, _jsxruntime.jsx)(_card_content.CardContent, {
                            title: title,
                            description: description,
                            preTitle: preTitle,
                            titleId: titleId,
                            descriptionId: descriptionId,
                            content: content
                        }),
                        decoratorsBoundary === 'card' && decorators
                    ]
                })
            }));
}));
const WithCardState = (0, _mobxreactlite.observer)(function WithCardState({ children }) {
    const state = {
        isHovered: (0, _card_context.useCardHover)(),
        isFocused: (0, _card_context.useCardFocus)(),
        isMouseDown: (0, _card_context.useCardMouseDown)()
    };
    return children(state);
});
function getBoundaries({ decoratorsBoundary, selectableBoundary, hoverBoundary }) {
    const fallbackBoundary = decoratorsBoundary ?? selectableBoundary ?? hoverBoundary ?? 'thumbnail';
    return {
        decoratorsBoundary: decoratorsBoundary ?? fallbackBoundary,
        selectableBoundary: selectableBoundary ?? fallbackBoundary,
        hoverBoundary: hoverBoundary ?? fallbackBoundary
    };
}
const getCardContainerAriaProps = ({ ariaLabelledBy, ariaLabel, cardContainerIsGroup, contentIds })=>{
    if (!cardContainerIsGroup) return {};
    const props = {
        role: 'group',
        ariaDescribedBy: contentIds?.descriptionId
    };
    if (ariaLabelledBy != null) return {
        ...props,
        ariaLabelledBy
    };
    if (ariaLabel != null) return {
        ...props,
        ariaLabel
    };
    return {
        ...props,
        ariaLabelledBy: contentIds?.titleId
    };
};
const getButtonAriaProps = ({ ariaLabelledBy, ariaLabel, buttonAriaLabelledBy, buttonAriaLabel, cardContainerIsGroup, contentIds })=>{
    const props = {
        buttonAriaDescribedBy: cardContainerIsGroup ? undefined : contentIds?.descriptionId
    };
    if (buttonAriaLabelledBy != null) return {
        ...props,
        buttonAriaLabelledBy
    };
    if (buttonAriaLabel != null) return {
        ...props,
        buttonAriaLabel
    };
    if (ariaLabelledBy != null) return {
        ...props,
        buttonAriaLabelledBy: ariaLabelledBy
    };
    if (ariaLabel != null) return {
        ...props,
        buttonAriaLabel: ariaLabel
    };
    return {
        ...props,
        buttonAriaLabelledBy: contentIds?.titleId
    };
};
const InternalCardButton = (0, _mobxreactlite.observer)(function InternalCardButton({ ariaControls, ariaHasPopup, ariaExpanded, ariaPressed, buttonAriaLabelledBy, buttonAriaDescribedBy, href, target, buttonAriaLabel, draggable, disabled, onDragStart, onDragEnd, buttonId, buttonRef, onContextMenu, tabIndex, onClick, onPointerDown, onPointerUp, onPointerEnter, onPointerLeave, onPointerMove, onPointerCancel }) {
    const { onFocus, onBlur } = (0, _card_context.useCardEventHandlers)();
    const preventDefault = _react.useCallback((e)=>e.preventDefault(), []);
    const doNothing = _react.useCallback(()=>{}, []);
    if (!(href || onClick)) return null;
    if (href) return ((0, _jsxruntime.jsx)("a", {
            id: buttonId,
            href: href,
            target: target,
            className: (0, _classnames.default)(_cardcss.default.cardButton, {
                [_cardcss.default.hideOutline]: _access_mode_state.accessModeState.isMouseMode
            }),
            onClick: disabled ? preventDefault : onClick,
            onPointerDown: disabled ? preventDefault : onPointerDown,
            onPointerUp: disabled ? preventDefault : onPointerUp,
            onPointerEnter: onPointerEnter,
            onPointerLeave: onPointerLeave,
            onPointerMove: onPointerMove,
            onPointerCancel: onPointerCancel,
            "aria-controls": ariaControls,
            "aria-haspopup": ariaHasPopup,
            "aria-labelledby": buttonAriaLabelledBy,
            "aria-describedby": buttonAriaDescribedBy,
            "aria-label": buttonAriaLabelledBy ? undefined : buttonAriaLabel,
            "aria-disabled": disabled || undefined,
            "aria-expanded": ariaHasPopup && ariaExpanded,
            draggable: draggable,
            onFocus: onFocus,
            onBlur: onBlur,
            onDragStart: onDragStart,
            onDragEnd: onDragEnd,
            onContextMenu: onContextMenu,
            ref: buttonRef,
            tabIndex: tabIndex
        }));
    const onKeyUp = (e)=>{
        if (!disabled && e.key === ' ') e.currentTarget.click();
    };
    const onKeyDown = (e)=>{
        if (!disabled) {
            if (e.key === ' ') e.preventDefault();
            else if (e.key === 'Enter') {
                e.preventDefault();
                e.currentTarget.click();
            }
        }
    };
    return (0, _jsxruntime.jsx)("div", {
        role: "button",
        "aria-disabled": disabled || undefined,
        "aria-controls": ariaControls,
        "aria-haspopup": ariaHasPopup,
        "aria-expanded": ariaExpanded,
        "aria-pressed": ariaPressed,
        "aria-label": buttonAriaLabelledBy ? undefined : buttonAriaLabel,
        "aria-labelledby": buttonAriaLabelledBy,
        "aria-describedby": buttonAriaDescribedBy,
        className: (0, _classnames.default)(_cardcss.default.divButton, _cardcss.default.cardButton, {
            [_cardcss.default.hideOutline]: _access_mode_state.accessModeState.isMouseMode
        }),
        draggable: draggable,
        id: buttonId,
        onClick: disabled ? doNothing : onClick,
        onPointerDown: disabled ? doNothing : onPointerDown,
        onPointerUp: disabled ? doNothing : onPointerUp,
        onPointerEnter: onPointerEnter,
        onPointerLeave: onPointerLeave,
        onPointerMove: onPointerMove,
        onPointerCancel: onPointerCancel,
        onContextMenu: onContextMenu,
        onFocus: onFocus,
        onBlur: onBlur,
        onKeyUp: onKeyUp,
        onKeyDown: onKeyDown,
        onDragStart: onDragStart,
        onDragEnd: onDragEnd,
        ref: buttonRef,
        tabIndex: tabIndex ?? 0
    });
});
const SelectableWrapper = (0, _mobxreactlite.observer)(({ children, boundary, borderRadius, className })=>{
    const { selected, selectableMode, selectableBoundary } = (0, _card_context.useCardConfiguration)();
    const enableAnimations = (0, _provider.useEnableAnimations)();
    const isInThumbnailContainerContext = (0, _thumbnail.useInThumbnailContainerContext)();
    if (!isInThumbnailContainerContext && selectableBoundary === boundary && selectableMode !== 'none') return (0, _jsxruntime.jsx)("div", {
        className: (0, _classnames.default)(_cardcss.default.selectableWrapper, className, {
            [_cardcss.default.selected]: selected,
            [_cardcss.default.selectableOnHover]: selectableMode === 'primary',
            [_cardcss.default.animated]: enableAnimations,
            [_cardcss.default.borderRadiusElement]: borderRadius === 'element',
            [_cardcss.default.borderRadiusElementSmall]: borderRadius === 'elementSmall',
            [_cardcss.default.borderRadiusElementRelaxed]: borderRadius === 'elementRelaxed',
            [_cardcss.default.borderRadiusElementSoftest]: borderRadius === 'elementSoftest',
            [_cardcss.default.borderRadiusElementRound]: borderRadius === 'elementRound'
        }),
        children: children
    });
    return children;
});
