// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/html_preview/html_preview.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "HtmlPreview", {
    enumerable: true,
    get: function() {
        return HtmlPreview;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = require("react");
const _html_previewmessages = require("./html_preview.messages");
function HtmlPreview({ html }) {
    const iframeRef = (0, _react.useRef)(null);
    const [heightPx, setHeightPx] = (0, _react.useState)(0);
    const resizeObserverRef = (0, _react.useRef)(null);
    const docHtml = (0, _react.useMemo)(()=>{
        const parser = new DOMParser();
        const doc = parser.parseFromString(html, 'text/html');
        doc.querySelectorAll('script').forEach((s)=>s.remove());
        doc.querySelectorAll('*').forEach((el)=>{
            for (const attr of Array.from(el.attributes))if (attr.name.toLowerCase().startsWith('on')) el.removeAttribute(attr.name);
        });
        return '<!doctype html>' + doc.documentElement.outerHTML;
    }, [
        html
    ]);
    const measureHeight = (0, _react.useCallback)(()=>{
        const iframe = iframeRef.current;
        if (!iframe) return;
        try {
            const doc = iframe.contentDocument;
            if (doc) {
                iframe.style.height = '0';
                const height = Math.max(doc.documentElement.scrollHeight, doc.body?.scrollHeight ?? 0);
                iframe.style.height = `${height}px`;
                setHeightPx(height);
            }
        } catch  {}
    }, []);
    const handleLoad = (0, _react.useCallback)(()=>{
        measureHeight();
        resizeObserverRef.current?.disconnect();
        const iframe = iframeRef.current;
        if (!iframe) return;
        try {
            const doc = iframe.contentDocument;
            if (doc?.documentElement) {
                resizeObserverRef.current = new ResizeObserver(measureHeight);
                resizeObserverRef.current.observe(doc.documentElement);
            }
        } catch  {}
    }, [
        measureHeight
    ]);
    (0, _react.useEffect)(()=>{
        return ()=>{
            resizeObserverRef.current?.disconnect();
        };
    }, []);
    return (0, _jsxruntime.jsx)("iframe", {
        ref: iframeRef,
        title: _html_previewmessages.HtmlPreviewMessages.title(),
        sandbox: "allow-same-origin",
        srcDoc: docHtml,
        onLoad: handleLoad,
        style: {
            width: '100%',
            height: `${heightPx}px`,
            border: 0,
            display: 'block'
        }
    });
}
