// @canva/app-ui-kit/dist/cjs/ui/base/surface/internal/content_container.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ContentContainer", {
    enumerable: true,
    get: function() {
        return ContentContainer;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _focus_lock = require('../../a11y/focus_lock/focus_lock');
function ContentContainer({ ariaLabel, ariaLabelledBy, ariaDescribedBy, captureFocus = true, autoFocus = true, role, onRequestEscape, children, ref,
className, ...rest }) {
    const divProps = {
        ['aria-label']: ariaLabel,
        ['aria-labelledby']: ariaLabelledBy,
        ['aria-describedby']: ariaDescribedBy,
        role,
        ...rest
    };
    const effectiveCaptureFocus = role === 'dialog' || role === 'alertdialog' ? true : captureFocus;
    return (0, _jsxruntime.jsx)(_focus_lock.FocusLock, {
        returnFocus: effectiveCaptureFocus,
        preventScroll: "on-autofocus",
        disabled: !effectiveCaptureFocus,
        onRequestEscape: onRequestEscape,
        autoFocus: autoFocus,
        ref: ref,
        lockProps: divProps,
        className: className,
        makeContainerFocusable: true,
        children: children
    });
}
