// @canva/app-ui-kit/dist/cjs/ui/base/color_picker/color_picker.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BaseColorPicker () {
        return _color_picker.BaseColorPicker;
    },
    get ColorPicker () {
        return _color_picker.ColorPicker;
    },
    get ColorPickerFlyout () {
        return _color_picker_flyout.ColorPickerFlyout;
    },
    get StatelessColorPickerFlyout () {
        return _color_picker_flyout.StatelessColorPickerFlyout;
    },
    get useColorPickerAlpha () {
        return _color_picker_provider.useColorPickerAlpha;
    },
    get useColorPickerColor () {
        return _color_picker_provider.useColorPickerColor;
    },
    get useColorPickerControls () {
        return _color_picker_provider.useColorPickerControls;
    },
    get useColorPickerHex () {
        return _color_picker_provider.useColorPickerHex;
    },
    get useColorPickerHue () {
        return _color_picker_provider.useColorPickerHue;
    },
    get useColorPickerSaturation () {
        return _color_picker_provider.useColorPickerSaturation;
    },
    get useColorPickerValue () {
        return _color_picker_provider.useColorPickerValue;
    }
});
const _color_picker = require("./private/color_picker");
const _color_picker_flyout = require("./private/color_picker_flyout");
const _color_picker_provider = require("./internal/color_picker_provider");
