// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/html_preview/html_preview.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import 'react';
import { HtmlPreviewMessages as Messages } from './html_preview.messages';
/** 
 * Renders HTML content in a sandboxed iframe with auto-sizing.
 * Canva Design Guidelines only recommend using this component in the previewUi of content publisher apps.
 * Scripts and inline event handlers are automatically stripped for security.
 * Uses allow-same-origin sandbox to measure content height from parent.
 */ export function HtmlPreview({ html }) {
    const iframeRef = useRef(null);
    const [heightPx, setHeightPx] = useState(0);
    const resizeObserverRef = useRef(null);
    const docHtml = useMemo(()=>{
        const parser = new DOMParser();
        const doc = parser.parseFromString(html, 'text/html');
        doc.querySelectorAll('script').forEach((s)=>s.remove());
        doc.querySelectorAll('*').forEach((el)=>{
            for (const attr of Array.from(el.attributes))if (attr.name.toLowerCase().startsWith('on')) el.removeAttribute(attr.name);
        });
        return '<!doctype html>' + doc.documentElement.outerHTML;
    }, [
        html
    ]);
    const measureHeight = useCallback(()=>{
        const iframe = iframeRef.current;
        if (!iframe) return;
        try {
            const doc = iframe.contentDocument;
            if (doc) {
                iframe.style.height = '0';
                const height = Math.max(doc.documentElement.scrollHeight, doc.body?.scrollHeight ?? 0);
                iframe.style.height = `${height}px`;
                setHeightPx(height);
            }
        } catch  {}
    }, []);
    const handleLoad = useCallback(()=>{
        measureHeight();
        resizeObserverRef.current?.disconnect();
        const iframe = iframeRef.current;
        if (!iframe) return;
        try {
            const doc = iframe.contentDocument;
            if (doc?.documentElement) {
                resizeObserverRef.current = new ResizeObserver(measureHeight);
                resizeObserverRef.current.observe(doc.documentElement);
            }
        } catch  {}
    }, [
        measureHeight
    ]);
    useEffect(()=>{
        return ()=>{
            resizeObserverRef.current?.disconnect();
        };
    }, []);
    return _jsx("iframe", {
        ref: iframeRef,
        title: Messages.title(),
        sandbox: "allow-same-origin",
        srcDoc: docHtml,
        onLoad: handleLoad,
        style: {
            width: '100%',
            height: `${heightPx}px`,
            border: 0,
            display: 'block'
        }
    });
}
