// @canva/app-ui-kit/dist/cjs/ui/base/responsive/private/breakpoint_store.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "breakpointStore", {
    enumerable: true,
    get: function() {
        return breakpointStore;
    }
});
const _debounce = require('../../../../base/debounce');
const _preconditions = require('../../../../base/preconditions');
const _metrics = require('../../metrics/metrics');
const _metricscss = _interop_require_default(require('../../metrics/metrics.css'));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const MEDIA_QUERIES = [
    _metricscss.default.smallUp,
    _metricscss.default.mediumUp,
    _metricscss.default.largeUp,
    _metricscss.default.xLargeUp
];
const BREAKPOINT_MAP = new Map([
    [
        0,
        _metrics.Breakpoint.DEFAULT
    ],
    [
        1,
        _metrics.Breakpoint.SMALL
    ],
    [
        2,
        _metrics.Breakpoint.MEDIUM
    ],
    [
        3,
        _metrics.Breakpoint.LARGE
    ],
    [
        4,
        _metrics.Breakpoint.XLARGE
    ]
]);
const DEFAULT_BREAKPOINT_FOR_TESTS = _metrics.Breakpoint.DEFAULT;
class BreakpointStore {
    setSsrBreakpoint(breakpoint) {
        this.serverBreakpoint = breakpoint;
    }
    setTestBreakpoint(breakpoint) {
        this.testBreakpoint = breakpoint;
        this.notifySubscribers();
    }
    resetConfiguration() {
        this.serverBreakpoint = undefined;
        this.testBreakpoint = undefined;
    }
    notifySubscribers() {
        this.subscribers.forEach((callback)=>callback());
    }
    constructor(){
        this.subscribers = new Set();
        this.subscribe = (callback)=>{
            this.subscribers.add(callback);
            return ()=>this.subscribers.delete(callback);
        };
        this.getSnapshot = ()=>this.clientBreakpoint ?? this.testBreakpoint ?? DEFAULT_BREAKPOINT_FOR_TESTS;
        this.getServerSnapshot = ()=>_preconditions.Preconditions.checkExists(this.serverBreakpoint, 'SSR breakpoint is not defined. Call setSsrBreakpoint() to configure a breakpoint on SSR pages.');
        this.update = ()=>{
            if (this.mqls) {
                const matchCount = this.mqls.filter((mql)=>mql.matches).length;
                this.clientBreakpoint = BREAKPOINT_MAP.get(matchCount);
                this.notifySubscribers();
            }
        };
        this.listener = (0, _debounce.debounce)(this.update, 16);
        const matchMedia = typeof window !== 'undefined' && window.matchMedia;
        if (matchMedia) {
            this.mqls = MEDIA_QUERIES.map((query)=>matchMedia(query));
            this.mqls.forEach((mql)=>safeAddListener(mql, this.listener));
            this.update();
        }
    }
}
const breakpointStore = new BreakpointStore();
function safeAddListener(mql, listener) {
    if (typeof mql.addEventListener === 'function') mql.addEventListener('change', listener);
    else mql.addListener(listener);
}
