// @canva/app-ui-kit/dist/cjs/ui/base/scroll/scrollable/private/scrollable.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Scrollable", {
    enumerable: true,
    get: function() {
        return Scrollable;
    }
});
const _jsxruntime = require("react/jsx-runtime");
require("react");
const _base_scrollable = require("./base_scrollable");
const _scrollable_with_indicator = require("./scrollable_with_indicator");
const Scrollable = (props)=>{
    const { children, direction = 'vertical', onScroll, onScrollStateChange, debounceMs, indicator, ariaLabel, role, ref } = props;
    if (indicator) return (0, _jsxruntime.jsx)(_scrollable_with_indicator.ScrollableWithIndicator, {
        direction: direction,
        onScroll: onScroll,
        onScrollStateChange: onScrollStateChange,
        debounceMs: debounceMs,
        ariaLabel: ariaLabel,
        role: role,
        ref: ref,
        type: indicator.type,
        background: indicator.background ? indicator.background : 'surface',
        children: children
    });
    return (0, _jsxruntime.jsx)(_base_scrollable.BaseScrollable, {
        direction: direction,
        onScroll: onScroll,
        onScrollStateChange: onScrollStateChange,
        debounceMs: debounceMs,
        ariaLabel: ariaLabel,
        role: role,
        ref: ref,
        children: children
    });
};
