// @canva/app-ui-kit/dist/cjs/ui/base/scroll/scrollable/private/scrollable_with_indicator.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ScrollableWithIndicator", {
    enumerable: true,
    get: function() {
        return ScrollableWithIndicator;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _debounce = require('../../../../../base/debounce');
const _preconditions = require('../../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _provider = require('../../../provider/provider');
const _base_scrollable = require("./base_scrollable");
const _scrollable_with_indicatorcss = _interop_require_default(require("./scrollable_with_indicator.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const isWebkit = ()=>typeof window !== 'undefined' ? /AppleWebKit\//i.test(window.navigator.userAgent) : undefined;
const runningInWebkit = isWebkit();
const ScrollableWithIndicator = (props)=>{
    const [showStartIndicator, setShowStartIndicator] = _react.useState(false);
    const [showEndIndicator, setShowEndIndicator] = _react.useState(false);
    const [widths, setWidths] = _react.useState(null);
    const isRtl = (0, _provider.useDirection)() === 'RTL';
    const { children, direction, onScroll, onScrollStateChange, debounceMs, ariaLabel, role, type, background, ref } = props;
    const applyWebkitRtlFix = runningInWebkit && isRtl && direction === 'horizontal';
    const enableAnimations = (0, _provider.useEnableAnimations)();
    const updateWidths = _react.useCallback((scrollState)=>{
        const { scrollable, containerWidth, scrollWidth } = scrollState;
        if (!scrollable || containerWidth === widths?.client && scrollWidth === widths?.scroll) return;
        setWidths({
            client: containerWidth,
            scroll: scrollWidth
        });
    }, [
        widths
    ]);
    const updateIndicators = _react.useCallback((scrollState)=>{
        const atStart = direction === 'horizontal' ? scrollState.atStart : scrollState.atTop;
        const atEnd = direction === 'horizontal' ? scrollState.atEnd : scrollState.atBottom;
        if (!scrollState.scrollable) {
            setShowStartIndicator(false);
            setShowEndIndicator(false);
        } else {
            setShowStartIndicator(!atStart);
            setShowEndIndicator(!atEnd);
        }
        if (applyWebkitRtlFix)
            updateWidths(scrollState);
    }, [
        direction,
        applyWebkitRtlFix,
        updateWidths
    ]);
    const handleScrollStateChange = _react.useMemo(()=>{
        const debouncedOnScrollStateChange = onScrollStateChange && debounceMs ? (0, _debounce.debounce)(onScrollStateChange, debounceMs, {
            leading: true
        }) : onScrollStateChange;
        return (scrollState)=>{
            updateIndicators(scrollState);
            debouncedOnScrollStateChange?.(scrollState);
        };
    }, [
        onScrollStateChange,
        debounceMs,
        updateIndicators
    ]);
    const handleOnScroll = _react.useMemo(()=>{
        return onScroll && debounceMs ? (0, _debounce.debounce)(onScroll, debounceMs, {
            leading: true
        }) : onScroll;
    }, [
        onScroll,
        debounceMs
    ]);
    const indicatorWrapperSharedStyles = (0, _classnames.default)(_scrollable_with_indicatorcss.default.indicatorWrapper, getDirectionClassName(direction));
    const indicatorSharedStyles = (0, _classnames.default)(_scrollable_with_indicatorcss.default.indicator, getBackgroundClassName(background), getIndicatorTypeClassName(type), getDirectionClassName(direction), {
        [_scrollable_with_indicatorcss.default.animate]: enableAnimations
    });
    return (0, _jsxruntime.jsxs)(_base_scrollable.BaseScrollable, {
        direction: direction,
        onScroll: handleOnScroll,
        onScrollStateChange: handleScrollStateChange,
        ariaLabel: ariaLabel,
        role: role,
        ref: ref,
        className: (0, _classnames.default)(_scrollable_with_indicatorcss.default.scrollContainer, getDirectionClassName(direction)),
        children: [
            (0, _jsxruntime.jsx)("div", {
                className: (0, _classnames.default)(indicatorWrapperSharedStyles, _scrollable_with_indicatorcss.default.start),
                style: {
                    right: applyWebkitRtlFix && !!widths ? `${widths.scroll - widths.client}px` : undefined
                },
                children: (0, _jsxruntime.jsx)("div", {
                    className: (0, _classnames.default)(indicatorSharedStyles, _scrollable_with_indicatorcss.default.start, {
                        [_scrollable_with_indicatorcss.default.isVisible]: showStartIndicator
                    })
                })
            }),
            (0, _jsxruntime.jsx)("div", {
                className: (0, _classnames.default)(indicatorWrapperSharedStyles, _scrollable_with_indicatorcss.default.end),
                style: {
                    right: applyWebkitRtlFix && !!widths ? `${widths.scroll}px` : undefined
                },
                children: (0, _jsxruntime.jsx)("div", {
                    className: (0, _classnames.default)(indicatorSharedStyles, _scrollable_with_indicatorcss.default.end, {
                        [_scrollable_with_indicatorcss.default.isVisible]: showEndIndicator
                    })
                })
            }),
            (0, _jsxruntime.jsx)("div", {
                className: _scrollable_with_indicatorcss.default.childrenWrapper,
                children: children
            })
        ]
    });
};
function getBackgroundClassName(background) {
    switch(background){
        case 'canvas':
            return _scrollable_with_indicatorcss.default.canvas;
        case 'tabdock':
            return _scrollable_with_indicatorcss.default.tabdock;
        case 'surface':
            return _scrollable_with_indicatorcss.default.surface;
        case 'page':
            return _scrollable_with_indicatorcss.default.page;
        default:
            throw new _preconditions.UnreachableError(background);
    }
}
function getIndicatorTypeClassName(type) {
    switch(type){
        case 'shadow':
            return _scrollable_with_indicatorcss.default.shadow;
        case 'fade':
            return _scrollable_with_indicatorcss.default.fade;
        default:
            throw new _preconditions.UnreachableError(type);
    }
}
function getDirectionClassName(direction) {
    switch(direction){
        case 'vertical':
            return _scrollable_with_indicatorcss.default.vertical;
        case 'horizontal':
            return _scrollable_with_indicatorcss.default.horizontal;
        default:
            throw new _preconditions.UnreachableError(direction);
    }
}
