// @canva/app-ui-kit/dist/cjs/ui/base/scroll/scrollable/private/base_scrollable.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BaseScrollable () {
        return BaseScrollable;
    },
    get useScrollable () {
        return useScrollable;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _debounce = require('../../../../../base/debounce');
const _compose_refs = require('../../../../../base/react/compose_refs');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _resizeobserverpolyfill = _interop_require_default(require("resize-observer-polyfill"));
const _supports_animation = require('../../../animation/supports_animation');
const _provider = require('../../../provider/provider');
const _scroll = _interop_require_wildcard(require('../../scroll'));
const _base_scrollablecss = _interop_require_default(require("./base_scrollable.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function useScrollable(options) {
    const { onScroll, onScrollStateChange, debounceMs = 0 } = options;
    const rtl = (0, _provider.useDirection)() === 'RTL';
    const ref = _react.useRef(null);
    const enableAnimations = (0, _provider.useEnableAnimations)();
    const prefersReducedMotion = (0, _supports_animation.usePrefersReducedMotion)();
    const disableEasing = !enableAnimations || prefersReducedMotion;
    const handle = _react.useMemo(()=>{
        return {
            getScrollState: ()=>{
                if (ref.current) return _scroll.getScrollState(ref.current, {
                    rtl
                });
            },
            scrollTo: (options)=>{
                if (ref.current) {
                    const easing = disableEasing ? 'instant' : options.easing || 'smooth';
                    _scroll.scrollTo(ref.current, {
                        rtl,
                        ...options,
                        easing
                    });
                }
            },
            scrollBy: (options)=>{
                if (ref.current) {
                    const easing = disableEasing ? 'instant' : options.easing || 'smooth';
                    _scroll.scrollBy(ref.current, {
                        rtl,
                        ...options,
                        easing
                    });
                }
            },
            scrollToPage: (options)=>{
                if (ref.current) {
                    const easing = disableEasing ? 'instant' : options.easing || 'smooth';
                    _scroll.scrollToPage(ref.current, {
                        rtl,
                        ...options,
                        easing
                    });
                }
            },
            scrollByPage: (options)=>{
                if (ref.current) {
                    const easing = disableEasing ? 'instant' : options.easing || 'smooth';
                    _scroll.scrollByPage(ref.current, {
                        rtl,
                        ...options,
                        easing
                    });
                }
            }
        };
    }, [
        rtl,
        disableEasing
    ]);
    const debouncedOnScroll = _react.useMemo(()=>{
        if (!onScroll && !onScrollStateChange) return;
        const onScrollWrapper = ()=>{
            const scrollState = handle.getScrollState();
            if (scrollState) {
                onScroll?.(scrollState);
                onScrollStateChange?.(scrollState);
            }
        };
        return debounceMs === 0 ? onScrollWrapper : (0, _debounce.debounce)(onScrollWrapper, debounceMs, {
            leading: true
        });
    }, [
        onScroll,
        onScrollStateChange,
        debounceMs,
        handle
    ]);
    const debouncedOnScrollStateChange = _react.useMemo(()=>{
        if (!onScrollStateChange) return;
        const onScrollStateChangeWrapper = ()=>{
            const scrollState = handle.getScrollState();
            if (scrollState) onScrollStateChange?.(scrollState);
        };
        return debounceMs === 0 ? onScrollStateChangeWrapper : (0, _debounce.debounce)(onScrollStateChangeWrapper, debounceMs, {
            leading: true
        });
    }, [
        onScrollStateChange,
        debounceMs,
        handle
    ]);
    _react.useEffect(()=>{
        if (!ref.current || !debouncedOnScrollStateChange) return;
        const resizeObserver = new _resizeobserverpolyfill.default(()=>debouncedOnScrollStateChange());
        resizeObserver.observe(ref.current);
        const mutationObserver = new MutationObserver(()=>debouncedOnScrollStateChange());
        mutationObserver.observe(ref.current, {
            childList: true,
            subtree: true
        });
        return ()=>{
            resizeObserver.disconnect();
            mutationObserver.disconnect();
        };
    }, [
        debouncedOnScrollStateChange
    ]);
    const setRef = _react.useCallback((element)=>{
        ref.current = element;
        if (element) debouncedOnScrollStateChange?.();
    }, [
        debouncedOnScrollStateChange
    ]);
    return {
        ref: setRef,
        handle,
        onScroll: debouncedOnScroll
    };
}
const BaseScrollable = (props)=>{
    const { children, direction, onScroll: onScrollProp, onScrollStateChange, debounceMs = 0, className, ariaLabel, role = 'region', ref: forwardedRef } = props;
    const { ref, handle, onScroll } = useScrollable({
        onScroll: onScrollProp,
        onScrollStateChange,
        debounceMs
    });
    const localRef = _react.useRef(null);
    _react.useImperativeHandle(forwardedRef, ()=>({
            ...handle,
            getLegacyScrollState: ()=>getLegacyScrollState(localRef.current, direction)
        }));
    return (0, _jsxruntime.jsx)("div", {
        tabIndex: 0,
        "aria-label": ariaLabel,
        role: ariaLabel ? role : undefined,
        className: (0, _classnames.default)(_base_scrollablecss.default.scrollContainer, className, {
            [_base_scrollablecss.default.vertical]: direction === 'vertical',
            [_base_scrollablecss.default.horizontal]: direction === 'horizontal'
        }),
        ref: (0, _compose_refs.composeRefs)(localRef, ref),
        onScroll: onScroll,
        children: children
    });
};
function getLegacyScrollState(container, direction) {
    const state = _scroll.getScrollState(container);
    if (!state.scrollable) return {
        scrollable: false
    };
    return {
        scrollable: state.scrollable,
        atStart: direction === 'vertical' ? state.atTop : state.atStart,
        atEnd: direction === 'vertical' ? state.atBottom : state.atEnd,
        raw: {
            scroll: {
                height: container.scrollHeight,
                width: container.scrollWidth,
                left: container.scrollLeft,
                top: container.scrollTop
            },
            client: {
                height: container.clientHeight,
                width: container.clientWidth,
                left: container.clientLeft,
                top: container.clientTop
            }
        }
    };
}
