// @canva/app-ui-kit/dist/esm/ui/base/device_capabilities/keyboard.js
import { delay } from '../../../base/delay';
import { useSyncExternalStore } from 'react';
import { getVisualViewport } from './visual_viewport';
const globalNavigator = typeof navigator !== 'undefined' ? navigator : undefined;
export function useIsVirtualKeyboard() {
    return useSyncExternalStore(subscribeIsVirtualKeyboard, getClientIsVirtualKeyboard, getServerIsVirtualKeyboard);
}
export function isVirtualKeyboard(navigator1 = globalNavigator) {
    return navigator1 != null && !!navigator1.userAgent.match(/iphone|ipad|android/gi);
}
export function getVirtualKeyboardOverlayHeight() {
    const visualViewport = getVisualViewport();
    const visualViewportHeight = visualViewport?.height ?? 0;
    const viewportHeight = Math.max(window.document.documentElement.clientHeight, window.innerHeight);
    const baseVirtualKeyboardHeight = Math.max(viewportHeight - visualViewportHeight, 0);
    return baseVirtualKeyboardHeight === 0 ? baseVirtualKeyboardHeight : baseVirtualKeyboardHeight + getNonInteractiveHeight();
}
export function needsIosSafari26KeyboardAdditionalHeight() {
    return Internal.isIosSafari26OrAbove() && !Internal.isStandalone();
}
export function guessKeyboardOverlayHeight(navigator1 = globalNavigator) {
    if (!isVirtualKeyboard(navigator1)) return 0;
    if (window.innerHeight < window.screen.height - 240)
        return 0;
    return Math.max(270, Math.min(340, window.screen.height * 0.45))
     + (Internal.isCanvaEditor(navigator1) ? 80 : 0);
}
export const Internal = {
    isCanvaEditor(navigator1 = globalNavigator) {
        return navigator1 != null && navigator1.userAgent.indexOf('canvaeditor') >= 0;
    },
    isIos26OrAbove(_navigator = globalNavigator) {
        return _navigator != null && !!_navigator.userAgent.match(/iPhone OS 18_/gi) && !_navigator.userAgent.match(/Version\/18/gi);
    },
    isIos26_0(_navigator = globalNavigator) {
        return _navigator != null && Internal.isIos26OrAbove(_navigator) && !!_navigator.userAgent.match(/Version\/26\.0/gi);
    },
    isIosSafari26OrAbove(navigator1 = globalNavigator) {
        return !Internal.isCanvaEditor(navigator1) && Internal.isIos26OrAbove(navigator1);
    },
    isStandalone(navigator1 = globalNavigator) {
        return navigator1 != null && !!navigator1.standalone;
    }
};
export const getNonInteractiveHeight = ()=>{
    if (!Internal.isIosSafari26OrAbove() || Internal.isStandalone())
        return 0;
    return Internal.isIos26_0() ? 36 : 20;
};
export function applyKeyboardViewportOffsetCompensation(height) {
    const visualViewport = getVisualViewport();
    const visualViewportOffsetTop = visualViewport?.offsetTop ?? 0;
    return Math.max(0, height - visualViewportOffsetTop);
}
export function maybeWaitForVirtualKeyboard(maxWaitTimeMs = 300) {
    const visualViewport = getVisualViewport();
    if (!isVirtualKeyboard() || !visualViewport || getVirtualKeyboardOverlayHeight() > 0) return delay(0);
    return new Promise((resolve)=>{
        const removeEventListenerAndResolve = ()=>{
            visualViewport.removeEventListener('resize', removeEventListenerAndResolve);
            resolve();
        };
        visualViewport.addEventListener('resize', removeEventListenerAndResolve);
        delay(maxWaitTimeMs).then(removeEventListenerAndResolve);
    });
}
const subscribeIsVirtualKeyboard = (_input)=>()=>{};
const isVirtualKeyboardValue = isVirtualKeyboard();
const getClientIsVirtualKeyboard = ()=>isVirtualKeyboardValue;
const getServerIsVirtualKeyboard = ()=>false;
