// @canva/app-ui-kit/dist/cjs/ui/base/surface/sheet/private/sheet_wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "SheetWrapper", {
    enumerable: true,
    get: function() {
        return SheetWrapper;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = _interop_require_wildcard(require("react"));
const _reacttransitiongroup = require("react-transition-group");
const _layer = require('../../../layer/layer');
const _provider = require('../../../provider/provider');
const _stable_function = require('../../../stable_function/stable_function');
const _sheetcss = _interop_require_default(require("./sheet.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function SheetWrapper(props) {
    const Wrapper = (0, _provider.useEnableAnimations)() ? TransitionWrapper : StaticWrapper;
    return (0, _jsxruntime.jsx)(Wrapper, {
        ...props
    });
}
const TRANSITION_DURATION = parseInt(_sheetcss.default.transitionTime, 10);
const TRANSITION_CLASS_NAMES = {
    appear: _sheetcss.default.transitioning,
    enter: _sheetcss.default.transitioning,
    exit: _sheetcss.default.transitioning,
    appearActive: _sheetcss.default.visible,
    appearDone: _sheetcss.default.visible,
    enterActive: _sheetcss.default.visible,
    enterDone: _sheetcss.default.visible
};
function TransitionWrapper(
    { open: propOpen, onOpenComplete, onCloseComplete, markOutsideInert, children }
) {
    const [transitionOpen, setTransitionOpen] = _react.useState(propOpen);
    const nodeRef = _react.useRef(null);
    return (0, _jsxruntime.jsx)(_layer.Layer, {
        open: propOpen || transitionOpen,
        markOutsideInert: markOutsideInert,
        children: (0, _jsxruntime.jsx)(_reacttransitiongroup.CSSTransition, {
            in: propOpen,
            appear: true,
            timeout: TRANSITION_DURATION,
            classNames: TRANSITION_CLASS_NAMES,
            onEnter: ()=>setTransitionOpen(true),
            onEntered: ()=>onOpenComplete?.(),
            onExited: ()=>{
                setTransitionOpen(false);
                onCloseComplete?.();
            },
            nodeRef: nodeRef,
            children: (0, _jsxruntime.jsx)("div", {
                ref: nodeRef,
                children: children
            })
        })
    });
}
function StaticWrapper({ open, onCloseComplete, onOpenComplete, markOutsideInert, children }) {
    const prevOpen = _react.useRef(false);
    const stableOnCloseComplete = (0, _stable_function.useStableFunction)(()=>onCloseComplete?.());
    const stableOnOpenComplete = (0, _stable_function.useStableFunction)(()=>onOpenComplete?.());
    _react.useEffect(()=>{
        if (prevOpen.current === open)
        return;
        prevOpen.current = open;
        if (open) stableOnOpenComplete();
        else stableOnCloseComplete();
        prevOpen.current = open;
    }, [
        open,
        stableOnOpenComplete,
        stableOnCloseComplete
    ]);
    return (0, _jsxruntime.jsx)(_layer.Layer, {
        open: open,
        markOutsideInert: markOutsideInert,
        children: (0, _jsxruntime.jsx)("div", {
            className: _sheetcss.default.visible,
            children: children
        })
    });
}
