// @canva/app-ui-kit/dist/cjs/ui/base/typography/internal/typography.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Display () {
        return Display;
    },
    get InheritColor () {
        return InheritColor;
    },
    get Text () {
        return Text;
    },
    get TextComponent () {
        return TextComponent;
    },
    get Title () {
        return Title;
    },
    get getDefaultHeadingTagName () {
        return getDefaultHeadingTagName;
    },
    get typographySizes () {
        return typographySizes;
    },
    get typographyTones () {
        return typographyTones;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _access_mode_state = require('../../a11y/access_mode_state/access_mode_state');
const _handle = require('../../handle/handle');
const _responsive = require('../../responsive/responsive');
const _typographycss = _interop_require_wildcard(require("./typography.css"));
const _typography_aria_attributes = require("./typography_aria_attributes");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const InheritColor = Symbol('inherit');
const typographyTones = [
    'primary',
    'secondary',
    'tertiary',
    'critical',
    'info',
    'success',
    'warn',
    InheritColor
];
const typographySizes = [
    'xxlarge',
    'xlarge',
    'large',
    'medium',
    'small',
    'xsmall',
    'xxsmall'
];
const Text = _react.memo(({ size = 'medium', tagName, weight, margins, ...props })=>{
    const internalTagName = tagName ?? 'p';
    const variant = weight ?? 'regular';
    const internalClassName = (0, _classnames.default)(getTextClassName({
        variant,
        size
    }), margins === 'legacy' && _typographycss.default.margin);
    return (0, _jsxruntime.jsx)(TextComponent, {
        ...props,
        tagName: internalTagName,
        internalClassName: internalClassName
    });
});
const Title = _react.memo(({ size = 'medium', tagName, margins, ...props })=>{
    const internalTagName = tagName ?? getDefaultHeadingTagName(size);
    const internalClassName = (0, _classnames.default)(getTextClassName({
        variant: 'title',
        size
    }), margins === 'legacy' && _typographycss.default.margin);
    return (0, _jsxruntime.jsx)(TextComponent, {
        ...props,
        tagName: internalTagName,
        internalClassName: internalClassName
    });
});
const Display = _react.memo(({ size, ...props })=>{
    const internalClassName = getTextClassName({
        variant: 'display',
        size
    });
    return (0, _jsxruntime.jsx)(TextComponent, {
        ...props,
        internalClassName: internalClassName
    });
});
function TextComponent(props) {
    const { id, testId, alignment = 'start', tone = 'primary', capitalization = 'default', tagName: TagName, lineClamp, wrapStyle, children, allowUserSelect, internalClassName, fontSizeAdjust, fontStackMode, className, elementTiming, ref, style, ...aria } = props;
    const hideOutline = (0, _access_mode_state.useIsMouseAccessMode)() && props.tabIndex === -1;
    const innerRef = _react.useRef(null);
    _react.useImperativeHandle(ref, ()=>(0, _handle.createBasicHTMLHandle)(innerRef));
    if (children === '' || children == null) return null;
    const textWrapClassName = getTextWrapClassName({
        lineClamp,
        wrapStyle
    });
    const lineClampStyles = lineClamp && lineClamp > 1 ? {
        WebkitLineClamp: lineClamp
    } : {};
    const finalClassName = (0, _classnames.default)(internalClassName, getFontStackClassName(fontStackMode ?? 'default'), getAlignmentClassNames(alignment), tone === InheritColor ? _typographycss.default.inheritColor : tone && (0, _typographycss.getStyle)(tone), textWrapClassName, allowUserSelect && _typographycss.default.allowUserSelect, capitalization === 'uppercase' && _typographycss.default.uppercase, hideOutline && _typographycss.default.hideOutline, className);
    return ((0, _jsxruntime.jsx)(TagName, {
                id: id,
                "data-testid": testId,
                className: finalClassName,
                style: {
                    ...style,
                    ...(fontSizeAdjust ? getFontSizeAdjustStyles(fontSizeAdjust) : {}),
                    ...lineClampStyles,
                    ...getAlignmentStyles(alignment)
                },
                tabIndex: props.tabIndex,
                elementtiming: elementTiming,
                ref: innerRef,
                ...(0, _typography_aria_attributes.getTypographyAriaAttributes)(aria),
                children: children
            }));
}
function getAlignmentClassNames(direction) {
    if (typeof direction === 'string') return [
        (0, _typographycss.getStyle)(direction)
    ];
    return [
        direction && _typographycss.default.hasAlignment
    ];
}
const alignmentPropertyMap = {
    default: _typographycss.customProperties.defaultAlignment,
    smallUp: _typographycss.customProperties.smallUpAlignment,
    mediumUp: _typographycss.customProperties.mediumUpAlignment,
    largeUp: _typographycss.customProperties.largeUpAlignment,
    xLargeUp: _typographycss.customProperties.xLargeUpAlignment
};
function getAlignmentStyles(alignment) {
    if (typeof alignment === 'string') return {};
    return {
        ...(alignment && (0, _responsive.generateStyle)(alignmentPropertyMap, alignment))
    };
}
function getTextWrapClassName({ lineClamp, wrapStyle }) {
    if (lineClamp == null && wrapStyle == null)
        return null;
    if (lineClamp === 1)
    return _typographycss.default.ellipsis;
    return {
        [_typographycss.default.lineClamp]: lineClamp && lineClamp > 1,
        [_typographycss.default.wrapStyleBalance]: wrapStyle === 'balance',
        [_typographycss.default.wrapStylePretty]: wrapStyle === 'pretty'
    };
}
function getVariantClassNamePart(variant) {
    switch(variant){
        case 'regular':
            return '';
        case 'title':
            return 'Title';
        case 'display':
            return 'Display';
        case 'bold':
            return 'Bold';
        default:
            throw new _preconditions.UnreachableError(variant);
    }
}
function getSizeClassNamePart(size) {
    switch(size){
        case 'xxlarge':
            return 'ExtraExtraLarge';
        case 'xlarge':
            return 'ExtraLarge';
        case 'large':
            return 'Large';
        case 'medium':
            return 'Medium';
        case 'small':
            return 'Small';
        case 'xsmall':
            return 'ExtraSmall';
        case 'xxsmall':
            return 'ExtraExtraSmall';
        default:
            throw new _preconditions.UnreachableError(size);
    }
}
function getTextClassName({ variant, size }) {
    return (0, _typographycss.getStyle)(`text${getVariantClassNamePart(variant)}${getSizeClassNamePart(size)}`);
}
function getDefaultHeadingTagName(size = 'medium') {
    switch(size){
        case 'xlarge':
            return 'h1';
        case 'large':
            return 'h2';
        case 'medium':
            return 'h3';
        case 'small':
            return 'h4';
        case 'xsmall':
            return 'h5';
        case 'xxsmall':
            return 'h6';
        default:
            throw new _preconditions.UnreachableError(size);
    }
}
function getFontStackClassName(fontStackMode) {
    switch(fontStackMode){
        case 'default':
            return undefined;
        case 'platform-agnostic':
            return _typographycss.default.platformAgnosticFonts;
        case 'fallback':
            return _typographycss.default.fallbackFonts;
        default:
            throw new _preconditions.UnreachableError(fontStackMode);
    }
}
function getFontSizeAdjustStyles(fontSizeAdjust) {
    return {
        fontSizeAdjust: `${fontSizeAdjust.fontMetric} ${fontSizeAdjust.fontHeightMultiple}`
    };
}
