// @canva/app-ui-kit/dist/esm/ui/base/color_picker/internal/handle.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { Box } from '../../box/box';
import { useIsomorphicLayoutEffect } from '../../hydration/use_isomorphic_layout_effect';
import { Pin } from '../../surface/pin/pin';
import { Swatch } from '../../swatch/swatch';
import styles from './handle.css';
export function Handle(
    { color, active = false, focused = false, preview = false, className, style }
) {
    const pinRef = React.useRef(null);
    useIsomorphicLayoutEffect(()=>{
        preview && pinRef.current?.update();
    });
    const shouldScale = focused || active;
    const handleClassName = classNames(className, styles.handle, {
        [styles.handleScaled]: shouldScale
    });
    const swatchClassName = classNames(styles.handleSwatch, {
        [styles.handleSwatchScaled]: shouldScale
    });
    return _jsx(Pin, {
        ref: pinRef,
        open: preview,
        placement: "top-center",
        reference: _jsx("div", {
            className: handleClassName,
            style: style,
            children: _jsx("div", {
                className: swatchClassName,
                style: {
                    backgroundColor: color
                }
            })
        }),
        children: _jsx(Box, {
            className: styles.tooltip,
            border: "low",
            borderRadius: "elementRound",
            background: "elevationSurfaceFloating",
            shadow: "elevationSurfaceFloating",
            position: "absolute",
            padding: "1u",
            children: _jsx(Swatch, {
                fill: [
                    color
                ],
                size: "xsmall"
            })
        })
    });
}
